// jamango/api.ts
import type { Mat4, Quat, Vec2, Vec3, Vec4 } from "./math/types";
import { Schema, SchemaType } from "./schema";
import { TraitMeta } from "./traits";
import type { GlbAnimationClipName } from "@jamango/content-client";
import type {
	AnimationPlaybackMode,
	CharacterAvatarConfig,
	CharacterMeshAttachmentInfo,
	CharacterMovementProperties,
	CharacterRigId,
	CharacterSlotId,
	CharacterVisualFacingMode,
	FirstPersonMeshAttachmentInfo,
	GlbMeshSpec,
	Impl,
	JsonSerialisable,
	LeaderboardEntry,
	LeaderboardId,
	MeshSetId,
	MotionType,
	PartialCharacterAvatarConfig,
	PlayerPermissions,
	SpecPropPhysics,
	TraitId,
	TraitTargetId,
	TraitType,
	UISettings,
	WorldBounds,
	WorldSettings,
	WorldSpawn,
} from "./types";
import {
	CollisionTraitFilters,
	EntityId,
	OnCharacterUseTraitFilters,
	ProjectileTraitFilters,
	TargetTraitFilters,
} from "./types";

declare const impl: Impl;

// Assets have been moved to assets.ts for better AI context efficiency.
// Import via: import { assets } from "jamango";


type Command<TData> = {
	name: string;
	_data: TData;
};

export const SCENE_NODE_TYPE_PROP: number = impl.SCENE_NODE_TYPE_PROP;
export const SCENE_NODE_TYPE_CHARACTER: number = impl.SCENE_NODE_TYPE_CHARACTER;
export const SCENE_NODE_TYPE_AREA: number = impl.SCENE_NODE_TYPE_AREA;

export const sceneTree = {
    "Piece": {
        "id": "a2091d28-9032-473b-acd1-a77764370902",
        "name": "Piece",
        "position": [
            -3.304,
            0,
            -3.26
        ],
        "type": 1
    },
    "Spawn": {
        "id": "abb85416-b7a0-4ab9-80a7-92772a468ae1",
        "name": "Spawn",
        "position": [
            15.5,
            0.5,
            7.5
        ],
        "type": 2
    },
    "Blank Card": {
        "id": "3db6209e-0318-4d1d-b80b-442ded8731f8",
        "name": "Blank Card",
        "position": [
            3,
            5,
            -17.5
        ],
        "type": 0
    },
    "Equippable Cards": {
        "id": "9216ea1a-5c6b-41da-9a62-19bc4ef20ccf",
        "name": "Equippable Cards",
        "position": [
            -1.5,
            0.5,
            -25.5
        ],
        "type": 2
    }
} as const;

/**
 * Indicates whether the current environment is a client or server.
 * 
 * Both isClient and isServer can be true at the same time if a game is running in peer-to-peer mode, not on a dedicated server.
 * 
 * ```ts
 * J.onGameStart(() => {
 *   if (J.net.isHost) {
 *     // logic that should only run on the host/server
 *     // ...
 *   }
 * });
 * 
 * J.onGameStart(() => {
 *   if (J.net.isClient) {
 *     // logic that should only run clients
 *     // ...
 *   }
 * });
 * ```
 * 
 * ### Networking Commands
 * 
 * #### `J.net.defineCommand<TData>(name: string) => Command<TData>`
 * - **Environment: both**
 * 
 * Defines a custom network command for sending data between clients and server. Returns a command object that can be used with send, sendToAll, and listen.
 * 
 * ```ts
 * // Define a command with TypeScript data
 * const MyCommand = J.net.defineCommand<{ value: number; message: string }>("myCommand");
 * ```
 * 
 * #### `J.net.send<TCommand>(command: TCommand, data: TCommand["_data"], target?: EntityId | EntityId[]) => void`
 * - **Environment: both**
 * 
 * Sends a command to specific targets (clients or server). If no target is specified, sends to the server.
 * 
 * ```ts
 * // Send to server from client
 * J.net.send(MyCommand, { value: 10, message: "Hello" });
 * 
 * // Send to specific player from server
 * J.net.send(MyCommand, { value: 20, message: "Response" }, playerId);
 * 
 * // Send to multiple players
 * J.net.send(MyCommand, { value: 30, message: "Broadcast" }, [playerId1, playerId2]);
 * ```
 * 
 * #### `J.net.sendToAll<TCommand>(command: TCommand, data: TCommand["_data"], except?: EntityId) => void`
 * - **Environment: both**
 * 
 * Sends a command to all clients except optionally one (usually the sender).
 * 
 * ```ts
 * // Send to all clients
 * J.net.sendToAll(MyCommand, { value: 100, message: "Everyone" });
 * 
 * // Send to all except one player
 * J.net.sendToAll(MyCommand, { value: 200, message: "Others" }, excludedPlayerId);
 * ```
 * 
 * #### `J.net.listen<TCommand>(command: TCommand, handler: (data: TCommand["_data"], playerId?: EntityId) => void) => () => void`
 * - **Environment: both**
 * 
 * Listens for a command and returns an unsubscribe function. The handler receives the data and optionally the sender's player ID.
 * 
 * ```ts
 * // Listen for a command
 * const unsubscribe = J.net.listen(MyCommand, (data, playerId) => {
 *   console.log("Received command:", data);
 *   if (playerId) {
 *     console.log("From player:", playerId);
 *   }
 * });
 * 
 * // Stop listening when done
 * unsubscribe();
 * ```
 * 
 * ## Scene Tree
*/
export const net: {
	isClient: boolean;
	isHost: boolean;
	defineCommand: <TData>(name: string) => Command<TData>;
	send: <TCommand extends Command<any>>(
		command: TCommand,
		data: TCommand["_data"],
		target?: EntityId | EntityId[],
	) => void;
	sendToAll: <TCommand extends Command<any>>(
		command: TCommand,
		data: TCommand["_data"],
		except?: EntityId,
	) => void;
	listen: <TCommand extends Command<any>>(
		command: TCommand,
		handler: (data: TCommand["_data"], playerId?: EntityId) => void,
	) => () => void;
} = impl.net;

/**
 * - **Environment: client and server**
 * 
 * Returns whether the chat is enabled in the world.
*/
export const getWorldSettings: () => WorldSettings = impl.getWorldSettings;

/**
 * - **Environment: client and server**
 * 
 * Sets world settings such as chat enabled/disabled, others in future.
 * 
 * ```ts
 * J.onGameStart(() => {
 *   J.setWorldSettings({ chatEnabled: false });
 * });
 * ```
*/
export const setWorldSettings: (settings: Partial<WorldSettings>) => void = impl.setWorldSettings;

/**
 * - **Environment: client**
 * 
 * Returns the current UI settings for the client.
*/
export const getUISettings: () => UISettings = impl.getUISettings;

/**
 * - **Environment: client**
 * 
 * Sets the UI settings for the client. You can pass a partial object to only update specific options.
 * 
 * ```ts
 * // hide all builtin UI elements
 * J.setUISettings({
 *     mobileCameraButton: false,
 *     mobileRightButtons: false,
 *     mobileInteractButton: false,
 *     mobileEmoteButton: false,
 * });
 * ```
*/
export const setUISettings: (settings: Partial<UISettings>) => void = impl.setUISettings;

/**
 * - **Environment: client**
 * 
 * Returns the entity ID of the local player when called on the client side.
*/
export const getLocalPlayer: () => EntityId = impl.getLocalPlayer;

/**
 * - **Environment: both**
 * 
 * Returns the current world time in seconds since the game started.
*/
export const getWorldTime: () => number = impl.getWorldTime;

/**
 * - **Environment: both**
 * 
 * Returns the world bounds that define the playable area. The bounds are an axis-aligned bounding box (AABB) in world coordinates.
*/
export const getWorldBounds: () => WorldBounds = impl.getWorldBounds;

/**
 * - **Environment: both**
 * 
 * Checks if a position is within the world bounds. Returns true if the position is inside the playable area, false otherwise.
 * 
 * ```ts
 * J.onGameTick((dt, t) => {
 *     // get the global world respawn position
 *     const spawn = J.getWorldSpawn();
 *     const entities = J.getAllCharacters();
 *     for (const id of entities) {
 *         // set any character back to spawn in case they fall out of bounds
 *         if (!J.isInWorldBounds(J.getEntityPosition(id))) {
 *             J.setEntityPosition(id, spawn.position, true);
 *         }
 *     }
 * });
 * ```
*/
export const isInWorldBounds: (pos: Vec3) => boolean = impl.isInWorldBounds;

/**
 * - **Environment: both**
 * 
 * Returns the world's spawn point configuration, including the spawn position and rotation angle.
*/
export const getWorldSpawn: () => WorldSpawn = impl.getWorldSpawn;

export const isTextureAtlasLoading: () => boolean = impl.isTextureAtlasLoading;

export const uiElement: HTMLDivElement | undefined = impl.uiElement;

/**
 * - **Environment: client**
 * Controls the crosshair display and spread offset. This function only works on the client side.
 * 
 * - `visible` - Controls whether the crosshair is visible or hidden. Default is true (visible).
 * - `spread` - Controls the crosshair spread (how far apart the crosshair lines are). Default is 0 (no spread).
 * 
 * ```ts
 * // Hide the crosshair
 * J.setCrosshair(false);
 * 
 * // Show crosshair with some spread
 * J.setCrosshair(true, 0.2);
 * ```
*/
export const setCrosshair: (visible?: boolean, spread?: number) => void = impl.setCrosshair;

/**
 * - **Environment: client**
 * 
 * Controls the block selector display. The block selector is the box that highlights which block you are aiming at. This function only works on the client side.
 * 
 * - `visible` - Controls whether the block selector is visible or hidden. Default is true (visible).
*/
export const setBlockSelector: (visible?: boolean) => void = impl.setBlockSelector;

export const BLOCK_SHAPE_CUBE: number = impl.BLOCK_SHAPE_CUBE;
export const BLOCK_SHAPE_AIR: number = impl.BLOCK_SHAPE_AIR;

/**
 * - **Environment: client**
 * 
 * Controls the block selector display. The block selector is the box that highlights which block you are aiming at. This function only works on the client side.
 * 
 * - `visible` - Controls whether the block selector is visible or hidden. Default is true (visible).
*/
export const setBlock: (position: Vec3, shape: number, type: string) => void = impl.setBlock;

/**
 * - **Environment: both**
 * 
 * Gets the shape of the block at the specified coordinates.
*/
export const getBlockShape: (position: Vec3) => number = impl.getBlockShape;

/**
 * - **Environment: both**
 * 
 * Gets the type of the block at the specified coordinates.
*/
export const getBlockType: (position: Vec3) => string = impl.getBlockType;

export const getSceneTreeNodeEntity: (sceneTreeNodeId: string) => EntityId | undefined =
	impl.getSceneTreeNodeEntity;

export const MOTION_TYPE_STATIC: number = impl.MOTION_TYPE_STATIC;
export const MOTION_TYPE_DYNAMIC: number = impl.MOTION_TYPE_DYNAMIC;
export const MOTION_TYPE_KINEMATIC: number = impl.MOTION_TYPE_KINEMATIC;

/**
 * - **Environment: both**
 * 
 * Gets the complete transformation matrix of an entity, combining position, rotation, and scale. Returns undefined if the entity doesn't exist.
 * 
 * ```ts
 * const matrix = J.getEntityMatrix(entityId);
 * if (matrix) {
 *   // Use matrix for advanced transformations
 *   const position = J.vec3.fromMat4Translation([0, 0, 0], matrix);
 *   const rotation = J.mat4.getRotation([0, 0, 0, 1], matrix);
 * }
 * ```
 * 
 * ---
 * 
 * ## Character mesh attachments
 * - **Environment: client** — not networked. Every client that should see the attachment must call these functions.
 * 
 * Attach a mesh to one of the character's body slots. Slots apply per-rig retargeting so the same offset works across all rigs (`defaultMale`, `gnome`, `ogre`, …).
 * 
 * Available slots: `"head" | "chest" | "waist" | "armLeft" | "armRight" | "handLeft" | "handRight" | "legLeft" | "legRight" | "footLeft" | "footRight"`
 * 
 * ### Full equip/unequip pattern
 * 
 * ```ts
 * const rot = J.quat.setAxisAngle(J.quat.create(), [1, 0, 0], -1.5);
 * let handle3p = -1;
 * let handleFp = -1;
 * 
 * function equip(characterId: J.EntityId) {
 *     handle3p = J.addCharacterMeshAttachmentGlb(characterId, { id: "item_smg" }, "handRight", [0, 0, 0], rot, 1.3);
 *     if (characterId === J.getLocalPlayer()) {
 *         handleFp = J.addFirstPersonMeshAttachmentGlb(characterId, { id: "item_smg" }, [0.5, -0.7, -0.7], [0, 0, 0, 1], 1);
 *     }
 *     J.characterPlayAnimation(characterId, "items_fightTwoHanded_idle_over", { loop: true, fadeIn: 0.2 });
 * }
 * 
 * function unequip(characterId: J.EntityId) {
 *     J.removeCharacterMeshAttachment(characterId, handle3p);
 *     if (handleFp >= 0) J.removeFirstPersonMeshAttachment(characterId, handleFp);
 *     J.characterStopAnimation(characterId, "items_fightTwoHanded_idle_over", { fadeOut: 0.2 });
 *     handle3p = -1;
 *     handleFp = -1;
 * }
 * ```
 * 
 * **Available item GLB ids:** `item_smg`, `item_sword`, `item_medShield` — any id from the `items_grp` in the character GLB. Avatar component ids (e.g. `nova_chest`, `aero_head`) also work.
 * 
 * ---
*/
export const getEntityMatrix: (id: EntityId) => Mat4 | undefined = impl.getEntityMatrix;

/**
 * - **Environment: both**
 * 
 * Gets the position of the entity with the given ID.
*/
export const getEntityPosition: (id: EntityId) => Vec3 | undefined = impl.getEntityPosition;

/**
 * - **Environment: both**
 * 
 * Sets the position of the entity with the given ID.
 * 
 * `predictable` (default `true`) describes whether this move is one clients already anticipate — normal, continuous per-tick movement — as opposed to a discontinuous jump. It affects both rendering and networking:
 * 
 * - **Client:** with `true` the mesh smoothly interpolates from its current rendered position to the new one; with `false` it snaps there instantly. A freshly `spawnProp`'d entity starts at the origin `(0,0,0)`, so placing it with the default `true` makes it visibly slide out of the origin over that tick's frames (more noticeable at high framerates) — pass `false` to snap.
 * - **Server:** with `true` the change replicates as an ordinary state update that clients smooth over and that a client predicting the entity may override with its own prediction; with `false` the server force-broadcasts a hard correction, so every client — including one predicting that entity — snaps to the authoritative transform.
 * 
 * In short: pass `false` for any discontinuous reposition (spawn placement, teleport, respawn), where clients must not interpolate or predict their way to the new position; leave the default `true` for continuous movement you drive each tick.
 * 
 * ```ts
 * const prop = J.spawnProp(visual);
 * J.setEntityPosition(prop, pos, false); // snap; the default `true` would lerp from (0,0,0)
 * ```
*/
export const setEntityPosition: (id: EntityId, position: Vec3, predictable?: boolean) => void =
	impl.setEntityPosition;

/**
 * - **Environment: both**
 * 
 * Gets the interpolated position (smoothed for rendering) of an entity. This differs from `getEntityPosition` as it includes client-side smoothing for visual continuity.
 * 
 * ```ts
 * const interpolatedPos = J.getEntityInterpolatedPosition(entityId);
 * if (interpolatedPos) {
 *   // Use for smooth visual effects
 *   J.spawnParticlesAtPosition(interpolatedPos, "particle#12345");
 * }
 * ```
*/
export const getEntityInterpolatedPosition: (id: EntityId) => Vec3 | undefined =
	impl.getEntityInterpolatedPosition;

/**
 * - **Environment: both**
 * 
 * Gets the rotation (quaternion) of the specified entity.
*/
export const getEntityQuaternion: (id: EntityId) => Quat | undefined = impl.getEntityQuaternion;

/**
 * - **Environment: both**
 * 
 * Sets the rotation (quaternion) of the specified entity. `predictable` behaves exactly as in `setEntityPosition`: the default `true` smoothly interpolates the rotation and replicates it as a predictable update, while `false` snaps the visual and force-broadcasts a hard correction from the server. Pass `false` for discontinuous rotations — e.g. orienting a freshly spawned prop (otherwise it slerps from its identity rotation) or applying a teleport's rotation.
*/
export const setEntityQuaternion: (id: EntityId, quaternion: Quat, predictable?: boolean) => void =
	impl.setEntityQuaternion;

/**
 * - **Environment: both**
 * 
 * Gets the interpolated quaternion (smoothed for rendering) of an entity. This provides smooth rotation values for visual effects.
 * 
 * ```ts
 * const interpolatedQuat = J.getEntityInterpolatedQuaternion(entityId);
 * if (interpolatedQuat) {
 *   // Use for smooth visual calculations
 *   const forward = J.vec3.transformQuat([0, 0, 1], [0, 0, 1], interpolatedQuat);
 * }
 * ```
*/
export const getEntityInterpolatedQuaternion: (id: EntityId) => Quat | undefined =
	impl.getEntityInterpolatedQuaternion;

/**
 * - **Environment: both**
 * 
 * Gets the current velocity of the entity.
*/
export const getEntityVelocity: (id: EntityId) => Vec3 | undefined = impl.getEntityVelocity;

/**
 * - **Environment: both**
 * 
 * Adds to the entity's current velocity.
*/
export const addEntityVelocity: (id: EntityId, velocity: Vec3, predictable?: boolean) => void =
	impl.addEntityVelocity;

/**
 * - **Environment: both**
 * 
 * Sets how much gravity affects an entity. 0 = no gravity, 1 = normal gravity, 2 = double gravity. Use this for anti-gravity zones, moon physics, or custom gravity effects.
 * 
 * ```ts
 * // Make an entity float (no gravity)
 * J.setEntityGravityFactor(propId, 0);
 * 
 * // Make an entity heavy (double gravity)
 * J.setEntityGravityFactor(propId, 2);
 * 
 * // Moon gravity (1/6 Earth gravity)
 * J.setEntityGravityFactor(propId, 0.166);
 * ```
*/
export const setEntityGravityFactor: (id: EntityId, gravityFactor: number) => void =
	impl.setEntityGravityFactor;

/**
 * - **Environment: both**
 * 
 * Sets the velocity of the entity.
*/
export const setEntityVelocity: (id: EntityId, velocity: Vec3, predictable?: boolean) => void =
	impl.setEntityVelocity;

/**
 * - **Environment: both**
 * 
 * Gets the angular velocity of the entity.
*/
export const getEntityAngularVelocity: (id: EntityId) => Vec3 | undefined = impl.getEntityAngularVelocity;

/**
 * - **Environment: both**
 * 
 * Adds to the entity's angular velocity.
*/
export const addEntityAngularVelocity: (id: EntityId, angularVelocity: Vec3, predictable?: boolean) => void =
	impl.addEntityAngularVelocity;

/**
 * - **Environment: both**
 * 
 * Sets the angular velocity of the entity.
*/
export const setEntityAngularVelocity: (id: EntityId, angularVelocity: Vec3, predictable?: boolean) => void =
	impl.setEntityAngularVelocity;

/**
 * - **Environment: both**
 * 
 * Sets the linear and angular velocity of a kinematic entity such that it reaches the target position and quaternion in the next physics tick.
 * 
 * This API is intended to be called every tick to set new velocities. If you stop calling this method and want an entity to stop moving entirely, you should call `setEntityVelocity` and `setEntityAngularVelocity` with zero vectors.
*/
export const moveKinematicEntity: (id: EntityId, targetPosition: Vec3, targetQuaternion: Quat) => void =
	impl.moveKinematicEntity;

/**
 * - **Environment: both**
 * 
 * Gets the scale of the entity.
*/
export const getEntityScale: (id: EntityId) => number | undefined = impl.getEntityScale;

/**
 * - **Environment: both**
 * 
 * Gets the positions of blocks that the entity is currently colliding with. Returns an empty array if the entity doesn't exist or is not a physical entity.
 * 
 * ```ts
 * // spawn a prop that can collide with blocks
 * const propId = J.spawnProp("prop#ABCDE12345");
 * 
 * // check which blocks the entity is colliding with
 * const collisions = J.getEntityBlockCollision(propId);
 * collisions.forEach((position) => {
 *   console.log(`Entity is colliding with block at (${position[0]}, ${position[1]}, ${position[2]})`);
 * 
 *   // change each colliding block to a red block
 *   J.setBlock(position, J.BLOCK_SHAPE_CUBE, "block#ABCDE12345");
 * });
 * ```
*/
export const getEntityBlockCollision: (id: EntityId) => Vec3[] = impl.getEntityBlockCollision;

/**
 * - **Environment: both**
 * 
 * Sets the scale of the entity.
*/
export const setEntityScale: (id: EntityId, scale: number) => void = impl.setEntityScale;

/**
 * - **Environment: client**
 * 
 * Sets the tint color for a prop or character entity. The tint is applied as a color overlay that is mixed with the final color based on the alpha component. Works on prop and character (creature/player) entities; other entity kinds are ignored. On characters and GLB-component props the tint color is quantized to 4 bits per channel (fine for flashes/overlays).
 * 
 * ```ts
 * // set a R = 50%, G = 30%, B = 20% tint, mixed in by 50%
 * J.setEntityTint(propId, [0.5, 0.3, 0.2, 0.5]);
 * 
 * // make an entity appear completely blue (full blue with 100% mix)
 * J.setEntityTint(entityId, [0.0, 0.0, 1.0, 1.0]);
 * 
 * // remove tinting (color is irrelevant, as it is mixed by 0%)
 * J.setEntityTint(entityId, [1.0, 1.0, 1.0, 0.0]);
 * ```
*/
export const setEntityTint: (id: EntityId, tint: Vec4) => void = impl.setEntityTint;

/**
 * - **Environment: client**
 * 
 * Sets the glow multiplier for a prop or character entity. The glow value is a float from 0 to 1 that controls how much the entity ignores darkness and shadows.
 * Changing this can allow your entity to visually stand out. Works on prop and character (creature/player) entities; other entity kinds are ignored. Default is 0.
 * 
 * ```ts
 * // make an entity have no glow (fully affected by lighting)
 * J.setEntityGlow(entityId, 0.0);
 * 
 * // make the entity glow fully (ignore lighting completely)
 * J.setEntityGlow(entityId, 1.0);
 * ```
*/
export const setEntityGlow: (id: EntityId, glow: number) => void = impl.setEntityGlow;

/**
 * - **Environment: client**
 * 
 * Sets the visibility of an entity. When set to false, the entity will be hidden from view. This is a client-side only concept - the entity still exists and functions normally on the server, but is not rendered on the client.
 * 
 * ```ts
 * // spawn entity
 * const propId = J.spawnProp(J.assets.props.MyCoin.id);
 * 
 * // hide entity
 * J.setEntityVisible(propId, false);
 * ```
*/
export const setEntityVisible: (id: EntityId, isVisible: boolean) => void = impl.setEntityVisible;

/**
 * - **Environment: client**
 * 
 * Gets the current visual offset applied to an entity. Returns `undefined` if the entity does not exist.
 * 
 * ```ts
 * const offset = J.getEntityVisualOffset(propId);
 * if (offset) {
 *   console.log(`Offset: x=${offset[0]}, y=${offset[1]}, z=${offset[2]}`);
 * }
 * ```
*/
export const getEntityVisualOffset: (id: EntityId) => Vec3 | undefined = impl.getEntityVisualOffset;

/**
 * - **Environment: client**
 * 
 * Applies a visual-only position offset to an entity's rendered position. This offset is purely cosmetic and does not affect physics, collisions, or networking. Perfect for juice effects like floating, bobbing, or screen shake.
 * 
 * ```ts
 * // make a prop float up and down
 * J.onGameRender(() => {
 *   const bobAmount = Math.sin(J.getWorldTime() * 0.003) * 0.2;
 *   J.setEntityVisualOffset(propId, [0, bobAmount, 0]);
 * });
 * 
 * // offset a character slightly to the right
 * J.setEntityVisualOffset(characterId, [0.5, 0, 0]);
 * 
 * // clear the offset
 * J.setEntityVisualOffset(propId, undefined);
 * ```
*/
export const setEntityVisualOffset: (id: EntityId, offset: Vec3) => void = impl.setEntityVisualOffset;

/**
 * - **Environment: client**
 * 
 * Gets the current visual quaternion applied to an entity. Returns `undefined` if the entity does not exist.
 * 
 * ```ts
 * const quat = J.getEntityVisualQuaternion(propId);
 * if (quat) {
 *   console.log(`Rotation: x=${quat[0]}, y=${quat[1]}, z=${quat[2]}, w=${quat[3]}`);
 * }
 * ```
*/
export const getEntityVisualQuaternion: (id: EntityId) => Quat | undefined = impl.getEntityVisualQuaternion;

/**
 * - **Environment: client**
 * 
 * Applies a visual-only rotation (as a quaternion) on top of the entity's actual rotation. This rotation is purely cosmetic and does not affect physics, collisions, or networking. Perfect for spin effects, wobbles, or other visual rotations.
 * 
 * The visual quaternion is multiplied with the entity's base rotation, so it's applied on top of the entity's actual orientation.
 * 
 * ```ts
 * // create a spinning effect (you'll need to update this each frame)
 * J.onGameRender(() => {
 *   const rotation = (J.getWorldTime() * 0.001) % (Math.PI * 2);
 *   const euler = new THREE.Euler(0, rotation, 0);
 *   const quat = new THREE.Quaternion().setFromEuler(euler);
 *   J.setEntityVisualQuaternion(propId, [quat.x, quat.y, quat.z, quat.w]);
 * });
 * 
 * // apply a 45-degree tilt
 * const euler = new THREE.Euler(0, 0, Math.PI / 4);
 * const quat = new THREE.Quaternion().setFromEuler(euler);
 * J.setEntityVisualQuaternion(propId, [quat.x, quat.y, quat.z, quat.w]);
 * 
 * // clear the visual rotation (identity quaternion)
 * J.setEntityVisualQuaternion(propId, [0, 0, 0, 1]);
 * ```
*/
export const setEntityVisualQuaternion: (id: EntityId, quaternion: Quat) => void =
	impl.setEntityVisualQuaternion;

/**
 * - **Environment: client**
 * 
 * Gets the current visual scale applied to an entity. Returns `undefined` if the entity does not exist.
 * 
 * ```ts
 * const scale = J.getEntityVisualScale(propId);
 * if (scale) {
 *   console.log(`Scale: x=${scale[0]}, y=${scale[1]}, z=${scale[2]}`);
 * }
 * ```
*/
export const getEntityVisualScale: (id: EntityId) => Vec3 | undefined = impl.getEntityVisualScale;

/**
 * - **Environment: client**
 * 
 * Applies a visual-only scale multiplier to an entity. The scale is applied component-wise (x, y, z) on top of the entity's actual scale. This is purely cosmetic and does not affect physics, collisions, or networking. Perfect for squash & stretch effects, impact animations, or other visual feedback.
 * 
 * ```ts
 * // make a prop appear 50% larger visually (without affecting physics)
 * J.setEntityVisualScale(propId, [1.5, 1.5, 1.5]);
 * 
 * // non-uniform scale for a pancake effect
 * J.setEntityVisualScale(propId, [1.5, 0.2, 1.5]);
 * 
 * // clear the visual scale
 * J.setEntityVisualScale(propId, [1, 1, 1]);
 * ```
*/
export const setEntityVisualScale: (id: EntityId, scale: Vec3) => void = impl.setEntityVisualScale;

/**
 * - **Environment: both**
 * 
 * Makes an entity look at a specific position. For character entities, uses spherical coordinates for camera control. For other entities, rotates them to face the target point.
 * 
 * ```ts
 * // spawn entities
 * const characterId = J.spawnCharacter("avatar#ABCDE12345");
 * const propId = J.spawnProp("prop#ABCDE12345");
 * 
 * // make the character look at a specific position (uses camera control)
 * J.entityLookAt(characterId, [10, 5, 20]);
 * 
 * // make a prop face a specific position (rotates the prop)
 * J.entityLookAt(propId, [0, 0, 10]);
 * 
 * // make an entity look at another entity's position
 * const targetPos = J.getEntityPosition(targetId);
 * if (targetPos) J.entityLookAt(characterId, targetPos);
 * ```
*/
export const entityLookAt: (id: EntityId, target: Vec3) => void = impl.entityLookAt;

/**
 * - **Environment: both**
 * 
 * Gets a player's permissions.
*/
export const getPlayerPermissions: (playerId: EntityId) => PlayerPermissions | undefined =
	impl.getPlayerPermissions;

/**
 * - **Environment: both**
 * 
 * Sets a player's permissions.
 * 
 * ```ts
 * // player setup
 * J.onPlayerJoin((playerId) => {
 *   // set permissions
 *   J.setPlayerPermissions(playerId, {
 *     canFly: false,
 *     canForceRespawn: false,
 *     canUseIndividualBlocks: false,
 *     canInteract: true,
 *   });
 * });
 * ```
*/
export const setPlayerPermissions: (playerId: EntityId, permissions: Partial<PlayerPermissions>) => void =
	impl.setPlayerPermissions;

/**
 * Attach a mesh at one of the character's slots. Two surfaces, both client-only and
 * never networked — callers must add on every client that should see the attachment.
 *
 * Slots use the same per-rig retargeting body parts use, so a single offset works
 * across every rig (defaultMale, gnome, ogre, …) without per-rig tuning.
 *
 * - `addCharacterMeshAttachmentProp(id, …)` — `id` resolves to a block-structure
 *   renderable: prop asset PK (`prop#…`, block-structure prop), block-structure asset PK
 *   (`blockStructure#…`), or single block ID (`block#…`).
 * - `addCharacterMeshAttachmentGlb(spec, …)` — inline `GlbMeshSpec` ({ id, colors? }).
 *   Covers character mesh components (e.g. `aero_chest`) and any items_grp id, with
 *   optional per-attachment color overrides.
 *
 * Slot ids: `head`, `chest`, `waist`, `armLeft`, `armRight`, `handLeft`, `handRight`,
 * `legLeft`, `legRight`, `footLeft`, `footRight`.
 *
 * Position/quaternion are local to the slot (world meters / radians). Scale accepts a
 * uniform number or per-axis Vec3. Returns a handle for `set` / `remove`, or -1 on failure.
 */
export const addCharacterMeshAttachmentProp: (
	characterId: EntityId,
	id: string,
	slot: CharacterSlotId,
	position: Vec3,
	quaternion: Quat,
	scale: Vec3 | number,
) => number = impl.addCharacterMeshAttachmentProp;

/**
 * Attach a GLB mesh at a character body slot (third-person, seen by all). Client-only, never networked.
 * @example J.addCharacterMeshAttachmentGlb(player, { id: "item_sword" }, "handRight", [0, 0, 0], rot, 1)
 * @returns handle for setCharacterMeshAttachment / removeCharacterMeshAttachment, or -1 on failure
 */
export const addCharacterMeshAttachmentGlb: (
	characterId: EntityId,
	spec: GlbMeshSpec,
	slot: CharacterSlotId,
	position: Vec3,
	quaternion: Quat,
	scale: Vec3 | number,
) => number = impl.addCharacterMeshAttachmentGlb;

/**
 * Update an existing attachment's transform and/or slot. Pass `undefined` for any
 * field to leave it unchanged.
 */
export const setCharacterMeshAttachment: (
	characterId: EntityId,
	handle: number,
	position?: Vec3,
	quaternion?: Quat,
	scale?: Vec3 | number,
	slot?: CharacterSlotId,
) => void = impl.setCharacterMeshAttachment;

export const removeCharacterMeshAttachment: (characterId: EntityId, handle: number) => void =
	impl.removeCharacterMeshAttachment;

export const clearCharacterMeshAttachments: (characterId: EntityId) => void =
	impl.clearCharacterMeshAttachments;

export const getCharacterMeshAttachments: (characterId: EntityId) => CharacterMeshAttachmentInfo[] =
	impl.getCharacterMeshAttachments;

/**
 * Attach a mesh in the local first-person view of `characterId`. Renders only when
 * the local camera is currently first-person on that character; hidden otherwise.
 * Transforms are camera-relative. Client-only, never networked.
 *
 * Same Prop / Glb split as `addCharacterMeshAttachment*`:
 * - `addFirstPersonMeshAttachmentProp(id, …)` — prop / block-structure / block id.
 * - `addFirstPersonMeshAttachmentGlb(spec, …)` — inline `GlbMeshSpec`.
 */
export const addFirstPersonMeshAttachmentProp: (
	characterId: EntityId,
	id: string,
	position: Vec3,
	quaternion: Quat,
	scale: Vec3 | number,
) => number = impl.addFirstPersonMeshAttachmentProp;

/**
 * Attach a GLB mesh as a local first-person viewmodel (no slot, camera-relative, owner-only). Client-only, never networked.
 * @example J.addFirstPersonMeshAttachmentGlb(player, { id: "item_sword" }, [0.5, -0.4, -1], [0, 0, 0, 1], 1)
 * @returns handle for updateFirstPersonMeshAttachment / removeFirstPersonMeshAttachment, or -1 on failure
 */
export const addFirstPersonMeshAttachmentGlb: (
	characterId: EntityId,
	spec: GlbMeshSpec,
	position: Vec3,
	quaternion: Quat,
	scale: Vec3 | number,
) => number = impl.addFirstPersonMeshAttachmentGlb;

export const updateFirstPersonMeshAttachment: (
	characterId: EntityId,
	handle: number,
	position?: Vec3,
	quaternion?: Quat,
	scale?: Vec3 | number,
) => void = impl.updateFirstPersonMeshAttachment;

export const removeFirstPersonMeshAttachment: (characterId: EntityId, handle: number) => void =
	impl.removeFirstPersonMeshAttachment;

export const clearFirstPersonMeshAttachments: (characterId: EntityId) => void =
	impl.clearFirstPersonMeshAttachments;

export const getFirstPersonMeshAttachments: (
	characterId: EntityId,
) => FirstPersonMeshAttachmentInfo[] = impl.getFirstPersonMeshAttachments;

/**
 * - **Environment: client**
 * 
 * Resolves a slot-local matrix into world space using the same per-rig retargeting attachments use. Returns `undefined` if the character has no visual or the slot is not bound on the active rig.
 * 
 * ```ts
 * // muzzle is a 10cm forward offset on the right hand
 * const muzzleLocal = J.mat4.fromTranslation(J.mat4.create(), [0, 0, 0.1]);
 * const muzzleWorld = J.getCharacterSlotWorldMatrix(characterId, "handRight", muzzleLocal);
 * if (muzzleWorld) {
 *   const origin = J.vec3.fromMat4Translation([0, 0, 0], muzzleWorld);
 *   J.spawnProjectile("#ff0", 50, characterId, null, origin, aimDirection);
 * }
 * ```
*/
export const getCharacterSlotWorldMatrix: (
	characterId: EntityId,
	slot: CharacterSlotId,
	localMatrix: Mat4,
) => Mat4 | undefined = impl.getCharacterSlotWorldMatrix;

/**
 * - **Environment: client**
 * 
 * Resolves a camera-local matrix into world space. Returns `undefined` when the local view is not first-person on that character.
 * 
 * #### First-person muzzle position
 * 
 * ```ts
 * // Muzzle position for shell ejection / muzzle flash (first-person only)
 * const fpLocal = J.mat4.fromTranslation(J.mat4.create(), [0.5, -0.3, -1.8]);
 * J.mat4.rotateY(fpLocal, fpLocal, -Math.PI / 2);
 * 
 * const m = J.getCharacterFirstPersonWorldMatrix(characterId, fpLocal);
 * if (m) {
 *     const pos: J.Vec3 = [m[12], m[13], m[14]]; // world position from matrix column 3
 *     const flash = J.spawnProp(FLASH_ASSET);
 *     J.setEntityPosition(flash, pos, false);
 * }
 * ```
*/
export const getCharacterFirstPersonWorldMatrix: (
	characterId: EntityId,
	localMatrix: Mat4,
) => Mat4 | undefined = impl.getCharacterFirstPersonWorldMatrix;

/**
 * - **Environment: both**
 * 
 * Checks if a character is alive. Returns undefined if the character doesn't exist.
 * 
 * ```ts
 * const alive = J.getCharacterAlive(playerId);
 * if (alive === false) {
 *   // Character is dead, respawn them
 *   J.respawnCharacter(playerId);
 * }
 * ```
*/
export const getCharacterAlive: (id: EntityId) => boolean | undefined = impl.getCharacterAlive;

/**
 * - **Environment: both**
 * 
 * Sets whether a character is alive. Set to false to "kill" a character, true to revive them.
 * 
 * ```ts
 * // Kill a character
 * J.setCharacterAlive(playerId, false);
 * 
 * // Revive a character
 * J.setCharacterAlive(playerId, true);
 * ```
*/
export const setCharacterAlive: (id: EntityId, alive: boolean) => void = impl.setCharacterAlive;

/**
 * - **Environment: both**
 * 
 * Gets the movement properties for a character entity. Returns undefined if the entity doesn't exist or is not a character.
*/
export const getCharacterMovementProperties: (id: EntityId) => CharacterMovementProperties | undefined =
	impl.getCharacterMovementProperties;

/**
 * Sets the movement properties for a character entity.
 * 
 * ```ts
 * J.onPlayerJoin((playerId) => {
 *   // set custom movement properties for the character
 *   J.setCharacterMovementProperties(playerId, {
 *     walkSpeed: 6,
 *   });
 * });
 * ```
*/
export const setCharacterMovementProperties: (
	id: EntityId,
	properties: Partial<CharacterMovementProperties>,
) => void = impl.setCharacterMovementProperties;

/**
 * Play an animation clip on a character. The clip's mode (base / additive / override) is
 * read from its name — `_add` clips blend additively on top of other layers, `_over` clips
 * play concurrently at full weight, and everything else occupies the single base slot
 * (replacing whatever locomotion would otherwise pick). Calling with a clip already active
 * updates its options.
 *
 * opts:
 *   mode         — what the clip does when it reaches its end:
 *                    "loop" — restart from the beginning, indefinitely.
 *                    "once" — play through once, then clear the clip (base returns to
 *                             locomotion, additive/override layers stop).
 *                    "hold" — play through once, then freeze on the final frame until
 *                             stopped. Ideal for death poses, knockouts, etc.
 *                  If unset, defaults to "loop" for base/override and "once" for additive.
 *   speed        — playback rate multiplier (default 1).
 *   fadeIn       — crossfade duration when the clip starts (default 0.1 for base, 0 otherwise).
 *   cancelOnMove — drop this clip as soon as the character starts moving / falling / etc.
 *                  Useful for emote-style base animations.
 */
// Use `GlbAnimationClipName | (string & {})` so IDE autocomplete still surfaces the known
// GLB clip names, but any string (e.g. from schema/trait data) is accepted at the type level.
/**
 * - **Environment: client**
 * 
 * Plays a GLB animation clip on a character. The clip's playback mode is derived from the clip name:
 * 
 * - clips with no suffix are **base** clips — they occupy the character's single base slot, replacing whichever locomotion or emote clip would otherwise play. Playing a new base clip automatically clears any previous base clip.
 * - clips ending in `_add` are **additive** — they blend on top of the base (and other additives) as a delta from the bind pose. Useful for hit reactions, lean, impulses, etc.
 * - clips ending in `_over` are **override** — they play concurrently with the base at full weight. `_over` clips are authored with partial-bone tracks, so e.g. `twoHanded_idle_over` only affects arms while legs keep following locomotion.
 * 
 * Every clip name present in the GLB is available via `J.assets.animations.<name>.id` (locomotion, emotes, `_add`, and `_over` — anything the mixer can resolve).
 * 
 * Options:
 * - `mode` — what the clip does when it reaches its end:
 *   - `"loop"` — restart from the beginning, indefinitely.
 *   - `"once"` — play through once, then clear the clip. A base clip returns the base slot to locomotion; additive/override layers stop.
 *   - `"hold"` — play through once, then freeze on the final frame until the clip is explicitly stopped (`characterStopAnimation` / `clearCharacterAnimations`) or replaced. Ideal for death poses, knockouts, etc.
 * 
 *   Defaults to `"loop"` for base/override clips and `"once"` for additive clips.
 * - `speed` — playback rate multiplier (default `1`).
 * - `fadeIn` — crossfade-in duration when the clip starts. Defaults to `0.1s` for base clips, `0` for additive/override.
 * - `cancelOnMove` — when `true`, the animation is dropped as soon as the character starts moving / falling / crouching / flying. Useful for emote-style base clips so they self-clear when the player walks away.
 * 
 * ```ts
 * // client.ts. Animation playback runs only on the client, so call these from client-side code.
 * J.onPlayerJoin((playerId) => {
 *   // looping dance emote, auto-cancel when the player moves
 *   J.characterPlayAnimation(playerId, J.assets.animations.emotes_danceBasic.id, {
 *     cancelOnMove: true,
 *   });
 * 
 *   // one-shot hit reaction on top of whatever locomotion is playing
 *   J.characterPlayAnimation(playerId, "locomotion_default_hit_add");
 * 
 *   // persistent two-handed upper-body pose while holding a weapon
 *   J.characterPlayAnimation(playerId, "twoHanded_idle_over");
 * 
 *   // death animation that plays once and freezes on the final frame
 *   J.characterPlayAnimation(playerId, J.assets.animations.locomotion_default_death.id, {
 *     mode: "hold",
 *     speed: 0.8,
 *   });
 * });
 * ```
*/
export const characterPlayAnimation: (
	id: EntityId,
	clip: GlbAnimationClipName | (string & {}),
	opts?: { mode?: AnimationPlaybackMode; speed?: number; fadeIn?: number; cancelOnMove?: boolean },
) => void = impl.characterPlayAnimation;

/**
 * - **Environment: client**
 * 
 * Stops a specific animation clip on a character. Has no effect if the clip isn't currently playing. For base clips, stopping the clip returns the base slot to whatever locomotion resolves to next frame.
 * 
 * ```ts
 * J.characterStopAnimation(playerId, "twoHanded_idle_over", { fadeOut: 0.2 });
 * ```
*/
export const characterStopAnimation: (
	id: EntityId,
	clip: GlbAnimationClipName | (string & {}),
	opts?: { fadeOut?: number },
) => void = impl.characterStopAnimation;

/** Stop every animation currently playing on the character (base, additive, and override). */
/**
 * - **Environment: client**
 * 
 * Stops every animation currently playing on the character — base, additive, and override. Locomotion takes over the base slot again on the next frame.
 * 
 * ```ts
 * J.clearCharacterAnimations(playerId);
 * ```
*/
export const clearCharacterAnimations: (id: EntityId) => void = impl.clearCharacterAnimations;

/**
 * - **Environment: both**
 * 
 * Gets the current view ray of the character.
*/
export const getCharacterViewRay: (id: number) =>
	| {
			origin: Vec3;
			direction: Vec3;
			hitEntityId: number | undefined;
			hitBlock: Vec3 | undefined;
	  }
	| undefined = impl.getCharacterViewRay;

/**
 * - **Environment: both**
 * 
 * Gets the input object for a character entity. The input object contains all current input states including movement, camera orientation, and action states.
 * 
 * ```ts
 * const playerId = J.getLocalPlayer();
 * const input = J.getCharacterInput(playerId)!;
 * console.log("Player input:", input);
 * // access input properties like input.nipple, input.camera, etc.
 * 
 * ```
*/
export const getCharacterInput: (id: EntityId) =>
	| {
			pointerLocked: boolean;
			nipple: { x: number; y: number };
			camera: { radius: number; phi: number; theta: number };
			isFlyingDown: boolean;
			isFlyingUp: boolean;
			isSprinting: boolean;
			isCrouching: boolean;
			isJumping: boolean;
			isPrimaryDown: boolean;
			isSecondaryDown: boolean;
	  }
	| undefined = impl.getCharacterInput;

/**
 * - **Environment: client**
 * 
 * Returns the screen-space pointer state — the mouse on desktop, the primary touch on mobile: `{ x, y, isDown, justPressed, justReleased, locked, touch }`. Positions are CSS pixels with a top-left origin, the same coordinate space as the `onGameUiCanvas` context and `getScreenPosition`. `justPressed` and `justReleased` are true only on the frame the press or release happened, and on a release frame `x`/`y` still hold the release position, so release-based click checks work. While `locked` (pointer lock, i.e. gameplay mouselook) there is no cursor, so skip hit-testing. On mobile `touch` is true and `x`/`y` are `-1, -1` while no touch is active.
 * 
 * This is primarily relevant for interactive canvas UI drawn with `onGameUiCanvas` — poll it every frame and hit-test your rects against it. HTML-based UI usually doesn't need it: elements under `uiElement` receive normal DOM pointer events directly.
 * 
 * ```ts
 * J.onGameUiCanvas((ctx, width, height) => {
 *   const p = J.getPointer();
 *   const x = width - 120, y = height - 60, w = 100, h = 40;
 *   const hover = !p.locked && p.x >= x && p.x < x + w && p.y >= y && p.y < y + h;
 *   ctx.fillStyle = hover ? "#3d8f27" : "#2e6e1c";
 *   ctx.fillRect(x, y, w, h);
 *   if (hover && p.justPressed) buyItem();
 * });
 * ```
*/
export const getPointer: () => {
	x: number;
	y: number;
	isDown: boolean;
	justPressed: boolean;
	justReleased: boolean;
	locked: boolean;
	touch: boolean;
} = impl.getPointer;

/**
 * Control pointer lock (gameplay mouselook). Client-only.
 *   "auto"     — the default: a click on the game canvas grabs pointer lock (built-in click-to-lock).
 *   "unlocked" — release pointer lock now and suppress the automatic re-lock, so clicks no longer
 *                grab lock and mouse movement no longer drives the camera. `getPointer().locked` and
 *                `getCharacterInput().pointerLocked` stay false while the cursor keeps updating. Use
 *                this for interactive cursor UI; call `setPointerLock("auto")` to restore gameplay.
 *   "locked"   — grab pointer lock now and resume normal mouselook.
 * On touch devices there is no cursor to lock, but the states still control input routing: "unlocked"
 * stops finger drags steering the camera/character and lets `getPointer()` feed the UI (locked === false),
 * while "auto"/"locked" give the camera ownership of touch again.
 */
/**
 * - **Environment: client**
 * 
 * Controls pointer lock (gameplay mouselook). `"auto"` is the default: a click on the game canvas grabs pointer lock. `"unlocked"` releases lock now and suppresses the automatic re-lock, so clicks no longer grab lock and mouse movement no longer drives the camera — `getPointer().locked` and `getCharacterInput().pointerLocked` stay false while the cursor keeps updating. `"locked"` grabs lock now and resumes normal mouselook. No-op on touch / mobile devices. Prefer this over `document.exitPointerLock()` / `element.requestPointerLock()` so the engine's own re-lock stays in sync. The state resets to `"auto"` on world load.
 * 
 * ```ts
 * // free the cursor for an interactive canvas shop, then restore gameplay
 * J.setPointerLock("unlocked");
 * J.setCursor("default");
 * // ...later, when the shop closes:
 * J.setPointerLock("auto");
 * ```
*/
export const setPointerLock: (state: "auto" | "unlocked" | "locked") => void = impl.setPointerLock;

/**
 * Set the CSS cursor shown over the game area ("default", "pointer", "grab", "grabbing", …). Client-only.
 * Safe to call every frame. No effect while the pointer is locked (the cursor is hidden during mouselook).
 */
/**
 * - **Environment: client**
 * 
 * Sets the CSS cursor shown over the game area (any valid cursor value — `"default"`, `"pointer"`, `"grab"`, `"grabbing"`, …). Safe to call every frame. No effect while the pointer is locked, since the cursor is hidden during mouselook.
 * 
 * ```ts
 * J.onGameUiCanvas((ctx, width, height) => {
 *   const p = J.getPointer();
 *   const hover = !p.locked && p.x < 120 && p.y < 40;
 *   J.setCursor(hover ? "pointer" : "default");
 * });
 * ```
*/
export const setCursor: (css: string) => void = impl.setCursor;

/**
 * - **Environment: both**
 * 
 * Enables or disables movement input processing for a character entity. When disabled (false), the character will not respond to movement input but other input (like camera) will still work. This is useful for taking control of player input to drive other entities like platforms or vehicles.
 * 
 * ```ts
 * 
 * // disable player movement input
 * J.setCharacterMovementInputEnabled(playerId, false)
 * 
 * // use player's movement input to control a platform
 * const input = J.getCharacterInput(playerId);
 * const platformSpeed = 5;
 * const velocity = [
 *   input.nipple.x * platformSpeed,  // left/right movement
 *   0,                               // no vertical movement
 *   -input.nipple.y * platformSpeed  // forward/back movement (inverted)
 * ];
 * J.setEntityVelocity(platformId, velocity);
 * 
 * ```
*/
export const setCharacterMovementInputEnabled: (id: EntityId, enabled: boolean) => void =
	impl.setCharacterMovementInputEnabled;

/**
 * - **Environment: client**
 * 
 * Sets how a character's rendered body decides which way to face. This is a local rendering preference and is never networked — each client controls it independently.
 * 
 * - `"movement"` (default) — the body faces the steering/movement direction, so strafing or back-pedaling turns the body toward where the character is actually moving.
 * - `"camera"` — the body faces the camera/look angle, staying aligned with where the character is aiming regardless of movement direction.
 * 
 * ```ts
 * 
 * // lock the body to face the aim direction (e.g. while drawing a bow or aiming down iron sights, scoping in)
 * J.setCharacterVisualFacingMode(playerId, "camera");
 * 
 * // restore the default steering-based facing
 * J.setCharacterVisualFacingMode(playerId, "movement");
 * 
 * ```
*/
export const setCharacterVisualFacingMode: (id: EntityId, mode: CharacterVisualFacingMode) => void =
	impl.setCharacterVisualFacingMode;

export const CHARACTER_RIG_IDS: readonly CharacterRigId[] = impl.CHARACTER_RIG_IDS;

export const CHARACTER_SLOT_IDS: readonly CharacterSlotId[] = impl.CHARACTER_SLOT_IDS;

export const MESH_SET_IDS: readonly MeshSetId[] = impl.MESH_SET_IDS;

/**
 * - **Environment: both**
 * 
 * Returns the photo url for a character entity ID, or undefined if the entity is not a character.
*/
export const getCharacterPhotoUrl: (entityId: EntityId) => string | undefined = impl.getCharacterPhotoUrl;

/**
 * - **Environment: both**
 * 
 * Returns the character's current avatar config (with all modifications), or `undefined` if the entity isn't a character.
 * 
 * ```ts
 * const avatar = J.getCharacterAvatar(playerId);
 * if (avatar) {
 *   console.log("Rig:", avatar.rigId);
 *   console.log("Skin:", avatar.skinColorPrimary);
 *   console.log("Hair:", avatar.hairColor);
 *   console.log("Head components:", avatar.components.head.map(c => c.id));
 * }
 * ```
*/
export const getCharacterAvatar: (entityId: EntityId) => CharacterAvatarConfig | undefined = impl.getCharacterAvatar;

/**
 * - **Environment: both**
 * 
 * Loads a pre-made avatar asset by ID and sets it as the character's avatar.
 * 
 * ```ts
 * J.setCharacterAvatarById(playerId, J.assets.avatars.CoolAvatar.id);
 * ```
*/
export const setCharacterAvatarById: (entityId: EntityId, avatarAssetId: string) => void =
	impl.setCharacterAvatarById;

/**
 * - **Environment: both**
*/
export const setCharacterAvatarByConfig: (entityId: EntityId, avatarConfig: CharacterAvatarConfig) => void =
	impl.setCharacterAvatarByConfig;

/**
 * - **Environment: both**
 * 
 * Returns the player's **original avatar**, the one stored when they joined the server, before any script modifications. Returns `undefined` if the entity isn't a player character. Use this to reset a player back to their base look.
 * 
 * ```ts
 * // Reset a player back to their base look after a costume swap.
 * const base = J.getPlayerBaseAvatar(playerId);
 * if (base) J.setCharacterAvatarByConfig(playerId, base);
 * ```
*/
export const getPlayerBaseAvatar: (entityId: EntityId) => CharacterAvatarConfig | undefined =
	impl.getPlayerBaseAvatar;

/**
 * - **Environment: both**
 * 
 * Layers a partial config on top of the character's current avatar.
 * 
 * 1. **Omitted = untouched.** Fields you don't pass stay as-is, at every level.
 * 2. **Components stack per slot.** `components.chest: [x]` adds `x` alongside what's already there (doesn't replace). Duplicate `id`s per slot are skipped.
 * 
 * Everything else (`rigId`, colors, face fields) simply overwrites when you pass it.
 * Animation overrides currently support `anim.idle`, `anim.run`, `anim.walk`, and `anim.crouch`; set any of them to a GLB clip name or `null` to use the default rig animation.
 * 
 * > Full replace? Use `setCharacterAvatarByConfig`.
 * 
 * ```ts
 * // Character starts with the player's base avatar (e.g. chest slot = [default_chest])
 * J.onPlayerJoin((playerId) => {
 *     J.addCharacterAvatarConfig(playerId, {
 *         hairColor: "#000000",                                                // overwrites hair
 *         face: { eyeId: "Eye_Editor.png" },                                   // overwrites just eyeId; other face fields stay
 *         components: {
 *             chest: [{ id: "aero_chest", colors: { "01": "#ffffff" } }],      // chest slot → [default_chest, aero_chest]
 *             legLeft: [{ id: "aero_legLeft", colors: { "01": "#fff333", "02": "#FFFFFF" } }],
 *             legRight: [{ id: "aero_legRight", colors: {} }],
 *         },
 *     });
 * });
 * ```
 * 
 * ```ts
 * // Zombie mode, change skin + face, components untouched
 * J.addCharacterAvatarConfig(playerId, {
 *     skinColorPrimary: "#00ff00",
 *     anim: { run: J.assets.animations.locomotion_zombie_run.id },
 *     face: { eyeId: "Eye_M_Void.png" },
 * });
 * 
 * // Change a player's locomotion style. Use J.assets.animations for autocomplete
 * // over every GLB animation clip available in the character rig.
 * J.addCharacterAvatarConfig(playerId, {
 *     anim: {
 *         idle: J.assets.animations.locomotion_happy_idle.id,
 *         run: J.assets.animations.locomotion_happy_run.id,
 *         walk: J.assets.animations.locomotion_happy_walk.id,
 *         crouch: J.assets.animations.locomotion_fight_crouch.id,
 *     },
 * });
 * 
 * // Undo all modifications, reset to the player's original avatar
 * const base = J.getPlayerBaseAvatar(playerId);
 * J.setCharacterAvatarByConfig(playerId, base);
 * ```
*/
export const addCharacterAvatarConfig: (
	entityId: EntityId,
	config: PartialCharacterAvatarConfig,
) => void = impl.addCharacterAvatarConfig;

/**
 * - **Environment: both**
 * 
 * Performs a raycast from the given origin in the specified direction for a certain distance. Returns an object with details about the hit.
 * 
 * ```ts
 * J.onGameStart(() => {
 *   const origin = [0, 0, 0]; // starting point of the ray
 *   const direction = [1, 0, 0]; // direction the ray should travel in
 *   const distance = 10; // the maximum distance the ray should travel
 *   const ignoreEntityIds = []; // optional array of entity IDs to ignore in the raycast, e.g. a player's own character
 * 
 *   const result = J.raycast(origin, direction, distance, ignoreEntityIds);
 * 
 *   console.log(
 *     "Raycast result:",
 *     result.hit,
 *     result.hitBlock,
 *     result.hitEntityId,
 *     result.hitDistance,
 *     result.hitPosition,
 *     result.hitNormal
 *   );
 * });
 * ```
*/
export const raycast: (
	origin: Vec3,
	direction: Vec3,
	distance: number,
	ignoreEntityIds?: EntityId[],
) => {
	hit: boolean;
	hitBlock: Vec3 | undefined;
	hitEntityId: EntityId | undefined;
	hitDistance: number;
	hitPosition: Vec3 | undefined;
	hitNormal: Vec3 | undefined;
} = impl.raycast;

export const voxelRaycast: (
	origin: Vec3,
	direction: Vec3,
	distance: number,
) => {
	hit: boolean;
	hitBlock: Vec3 | undefined;
	hitDistance: number;
	hitPosition: Vec3 | undefined;
	hitNormal: Vec3 | undefined;
	hitBlockShape: number;
	hitBlockType: number;
} = impl.voxelRaycast;

/**
 * - **Environment: both**
 * 
 * Returns the username for a player entity ID, or undefined if the entity is not a player or the entity is not in the world.
 * 
 * ```ts
 * J.onPlayerJoin((playerId) => {
 *   const username = J.getPlayerUsername(playerId)!;
 *   J.sendChatMessage(`Welcome to the game, ${username}!`);
 * });
 * ```
*/
export const getPlayerUsername: (playerId: EntityId) => string | undefined = impl.getPlayerUsername;

/**
 * - **Environment: both**
 * 
 * Returns the photo url for a player entity ID, or undefined if the entity is not a player or the entity is not in the world.
*/
export const getPlayerPhotoUrl: (playerId: EntityId) => string | undefined = impl.getPlayerPhotoUrl;

/**
 * - **Environment: client**
 * 
 * Sets the nameplate's options for a character entity ID (players or NPCs). Nameplates are visible by default. It only works on Client.
 * 
 * The `text` parameter allows overriding the displayed nameplate text. If `null` or not provided, players will show their peer username automatically, while NPCs will have no nameplate.
 * 
 * ```ts
 * // client.ts
 * J.onPlayerJoin((playerId) => {
 *   // enable nameplate and set green color
 *   J.setCharacterNameplate(playerId, true, [0, 1, 0]);
 * 
 *   // override player nameplate text
 *   J.setCharacterNameplate(playerId, true, undefined, "🎮 Player");
 * });
 * 
 * // Set custom nameplate for NPC
 * const guardId = J.spawnCharacter("guard_avatar");
 * J.setCharacterNameplate(guardId, true, [1, 1, 0], "Town Guard");
 * 
 * // Reset to default (peer username for players, hidden for NPCs)
 * J.setCharacterNameplate(characterId, true, undefined, null);
 * ```
*/
export const setCharacterNameplate: (
	characterId: EntityId,
	visible?: boolean,
	textColor?: Vec3,
	text?: string,
) => void = impl.setCharacterNameplate;

/**
 * - **Environment: both**
 * 
 * Returns the number of peers connected to the world.
 * 
 * ```ts
 * J.onPlayerJoin(() => {
 *   const peerCount = J.getWorldPeerCount();
 *   J.sendChatMessage(`player joined! ${peerCount} players in game!`);
 * });
 * ```
*/
export const getWorldPeerCount: () => number = impl.getWorldPeerCount;

/**
 * - **Environment: both**
 * 
 * Spawns a new character using the given asset.
*/
export const spawnCharacter: (avatarAssetId: string) => EntityId = impl.spawnCharacter;

/**
 * - **Environment: both**
 * 
 * Spawns a new prop using the given asset.
*/
export const spawnProp: (propAssetId: string) => EntityId = impl.spawnProp;

/**
 * Spawn a Prop entity from an inline GLB spec. Client-only; spec is not replicated.
 * TODO: no physics body until sphere/box allocation for non-block-structure props lands.
 */
export const spawnPropFromSpec: (spec: GlbMeshSpec, physics?: SpecPropPhysics) => EntityId =
	impl.spawnPropFromSpec;

/**
 * - **Environment: both**
 * 
 * Returns whether physics is enabled for the given entity. Returns undefined if the entity doesn't exist.
*/
export const getEntityPhysicsEnabled: (entityId: EntityId) => boolean | undefined =
	impl.getEntityPhysicsEnabled;

/**
 * - **Environment: both**
 * 
 * Enables or disables physics for the given entity. Supports props and characters.
 * 
 * Note that disabling physics will prevent it from being hit by raycasts or being detected by sensors. Use this when an entity should be entirely non-physical.
 * 
 * ```ts
 * // spawn a prop
 * const propId = J.spawnProp("prop#ABCDE12345");
 * 
 * // disable physics for the prop
 * J.setEntityPhysicsEnabled(propId, false);
 * ```
*/
export const setEntityPhysicsEnabled: (entityId: EntityId, enabled: boolean) => void =
	impl.setEntityPhysicsEnabled;

/**
 * - **Environment: both**
 * 
 * Returns the current prop asset for the given prop entity.
*/
export const getPropAsset: (entityId: EntityId) => string | undefined = impl.getPropAsset;

/**
 * - **Environment: both**
 * 
 * Updates the asset of an existing prop entity. Only works on entities that are props.
 * 
 * ```ts
 * // spawn a prop
 * const propId = J.spawnProp("prop#ABCDE12345");
 * 
 * // change the prop's asset to a different one
 * J.updatePropAsset(propId, "prop#FGHIJ67890");
 * 
 * // make the prop invisible by setting asset to undefined
 * J.updatePropAsset(propId, undefined);
 * ```
*/
export const updatePropAsset: (entityId: EntityId, assetId: string | undefined) => void =
	impl.updatePropAsset;

/**
 * - **Environment: both**
 * 
 * Updates the physics properties of an existing prop entity. Only works on entities that are props. All properties in the object are optional.
 * 
 * ```ts
 * // spawn a prop
 * const propId = J.spawnProp("prop#ABCDE12345");
 * 
 * // make the prop dynamic and bouncy
 * J.updatePropPhysicsProperties(propId, {
 *   motionType: J.MOTION_TYPE_DYNAMIC,
 *   sensor: false,
 *   scale: 1.0,
 *   mass: 5.0,
 *   friction: 0.1,
 *   restitution: 0.8,
 *   gravityFactor: 1.0,
 * });
 * 
 * // change motion type to dynamic (affected by gravity and collisions)
 * J.updatePropPhysicsProperties(propId, { motionType: J.MOTION_TYPE_DYNAMIC });
 * 
 * // change motion type to static (not affected by gravity, no collisions, no movement)
 * J.updatePropPhysicsProperties(propId, { motionType: J.MOTION_TYPE_STATIC });
 * 
 * // change motion type to kinematic (not affected by gravity or collisions, but can be moved with velocity)
 * J.updatePropPhysicsProperties(propId, { motionType: J.MOTION_TYPE_KINEMATIC });
 * 
 * // make it a sensor (no collision) and scale it up
 * J.updatePropPhysicsProperties(propId, { sensor: true, scale: 2.0 });
 * ```
*/
export const updatePropPhysicsProperties: (
	entityId: EntityId,
	properties: {
		motionType?: MotionType;
		sensor?: boolean;
		scale?: number;
		mass?: number;
		friction?: number;
		restitution?: number;
		gravityFactor?: number;
	},
) => void = impl.updatePropPhysicsProperties;

/**
 * - **Environment: both**
 * 
 * Gets the current physics properties of a prop entity. Returns undefined if the entity doesn't exist or is not a prop. The returned object is a shallow copy that can be modified and passed to `updatePropPhysicsProperties`.
 * 
 * ```ts
 * // spawn a prop
 * const propId = J.spawnProp("prop#ABCDE12345");
 * 
 * // get current physics properties
 * const properties = J.getPropPhysicsProperties(propId);
 * if (properties) {
 *   console.log("Current properties:", properties);
 * 
 *   // modify properties and update (get -> edit -> set pattern)
 *   properties.mass *= 2; // double the mass
 *   properties.friction = 0.5; // change friction
 *   J.updatePropPhysicsProperties(propId, properties);
 * 
 *   // or just update specific properties
 *   J.updatePropPhysicsProperties(propId, { mass: properties.mass * 2 });
 * }
 * ```
*/
export const getPropPhysicsProperties: (entityId: EntityId) =>
	| {
			motionType: MotionType;
			sensor: boolean;
			scale: number;
			mass: number;
			friction: number;
			restitution: number;
			gravityFactor: number;
	  }
	| undefined = impl.getPropPhysicsProperties;

/**
 * - **Environment: client**
 * 
 * Spawns a particle effect using the given asset.
*/
export const spawnParticles: (particlesAssetId: string) => EntityId = impl.spawnParticles;

/**
 * - **Environment: both**
 * 
 * Spawns a box-shaped area with the given dimensions.
*/
export const spawnBoxArea: (sizeX: number, sizeY: number, sizeZ: number) => EntityId = impl.spawnBoxArea;

/**
 * - **Environment: both**
 * 
 * Spawns a spherical area with the given radius.
*/
export const spawnSphereArea: (radius: number) => EntityId = impl.spawnSphereArea;

/**
 * - **Environment: both**
 * 
 * Removes the specified entity.
*/
export const removeEntity: (id: EntityId) => void = impl.removeEntity;

/**
 * - **Environment: both**
 * 
 * Gets an area entity's description, which includes its shape type and shape settings. Returns undefined if the entity is not an area entity.
 * 
 * ```ts
 * // spawn a box area
 * const areaId = J.spawnBoxArea(5, 5, 5);
 * 
 * // get the area entity description
 * const areaDesc = J.getAreaEntityDesc(areaId);
 * 
 * console.log("Area Description:", areaDesc); // { type: "box", extents: [5, 5, 5] }
 * ```
*/
export const getAreaEntityDesc: (
	id: EntityId,
) => undefined | { type: "box"; extents: Vec3 } | { type: "sphere"; radius: number } = impl.getAreaEntityDesc;

/**
 * - **Environment: both**
 * 
 * ```ts
 * // spawn a box area
 * const areaId = J.spawnBoxArea(5, 5, 5);
 * 
 * // later, resize the area to different dimensions
 * J.setAreaEntityDesc(areaId, = {
 *   type: "box",
 *   extents: [10, 10, 10],
 * });
 * 
 * // change the box area to a sphere with a radius of 12
 * J.setAreaEntityDesc(areaId, {
 *   type: "sphere",
 *   radius: 12,
 * });
 * ```
*/
export const setAreaEntityDesc: (
	id: EntityId,
	desc: { type: "box"; extents: Vec3 } | { type: "sphere"; radius: number },
) => void = impl.setAreaEntityDesc;

/**
 * - **Environment: both**
 * 
 * Returns all block positions within a spherical area centered at the given position.
 * 
 * ```ts
 * const blocks = J.getBlocksInSphere([0, 0, 0], 5);
 * blocks.forEach((position) => {
 *   J.setBlock(position, J.BLOCK_SHAPE_CUBE, "block#ABCDE12345");
 * });
 * ```
*/
export const getBlocksInSphere: (center: Vec3, radius: number) => Vec3[] = impl.getBlocksInSphere;

/**
 * - **Environment: both**
 * 
 * Returns all block positions within a box area between the min and max coordinates.
 * 
 * ```ts
 * const blocks = J.getBlocksInBox([-5, 0, -5], [5, 10, 5]);
 * blocks.forEach((position) => {
 *   J.setBlock(position, J.BLOCK_SHAPE_AIR, "block#ABCDE12345");
 * });
 * ```
*/
export const getBlocksInBox: (min: Vec3, max: Vec3) => Vec3[] = impl.getBlocksInBox;

export const getBlocksInAreaEntity: (areaEntityId: EntityId) => Vec3[] = impl.getBlocksInAreaEntity;

/**
 * - **Environment: both**
 * 
 * Gets all entities within a given spherical area.
*/
export const getEntitiesInSphere: (center: Vec3, radius: number) => EntityId[] = impl.getEntitiesInSphere;

/**
 * - **Environment: both**
 * 
 * Defines a new trait with optional properties. The trait can be registered only once with a specific name.
*/
export const defineTrait: <S extends Schema>(
	name: string,
	schema?: S,
	meta?: TraitMeta,
) => string & { _schema: S; _traitDataType: S extends undefined ? any : SchemaType<S> } = impl.defineTrait;

/**
 * - **Environment: both**
 * 
 * Sets a trait on a target entity or object.
*/
export const setTrait: <T extends TraitType>(
	target: TraitTargetId,
	trait: T,
	value?: T["_traitDataType"],
) => void = impl.setTrait;

/**
 * - **Environment: both**
 * 
 * Gets the value of a trait from a target.
*/
export const getTrait: <T extends TraitType>(target: TraitTargetId, trait: T) => T["_traitDataType"] | null =
	impl.getTrait;

/**
 * - **Environment: both**
 * 
 * Removes a trait from a target.
*/
export const removeTrait: <T extends TraitType>(target: TraitTargetId, trait: T) => void = impl.removeTrait;

type ExtractTraitTuple<T extends [TraitType, ...TraitType[]]> = {
	[K in keyof T]: T[K]["_traitDataType"];
};

/**
 * - **Environment: both**
 * 
 * Gets all entities with the specified traits.
*/
export const getAllWithTraits: <T extends [TraitType, ...TraitType[]]>(
	traits: T,
) => Array<[id: any, ...traits: ExtractTraitTuple<T>]> = impl.getAllWithTraits;

/**
 * - **Environment: client**
 * 
 * Plays a sound using the specified asset and options.
 * 
 * ```ts
 * // play a sound globally
 * J.playSound(J.assets.audio.MySound.id, {
 *   volume: 0.5,
 * });
 * 
 * // play looping background music
 * J.playSound(J.assets.audio.BackgroundMusic.id, {
 *   volume: 0.3,
 *   loop: true,
 * });
 * ```
*/
export const playSound: (
	audioAssetId: string,
	options?: { volume?: number; detune?: number; fadeInDuration?: number; loop?: boolean },
) => number | undefined = impl.playSound;

/**
 * - **Environment: client**
 * 
 * Plays a spatial sound at a given position using the specified asset and options.
 * 
 * ```ts
 * // play a sound at a specific position in the world
 * J.playSoundAtPosition(J.assets.audio.MySound.id, [10, 5, -10], {
 *   volume: 0.5,
 *   loop: false,
 * });
 * 
 * // play looping ambient noise at a specific position
 * J.playSoundAtPosition(J.assets.audio.AmbientNoise.id, [0, 0, 0], {
 *   volume: 0.2,
 *   loop: true,
 * });
 * ```
*/
export const playSoundAtPosition: (
	audioAssetId: string,
	position: Vec3,
	options?: { volume?: number; detune?: number; fadeInDuration?: number; loop?: boolean },
) => number | undefined = impl.playSoundAtPosition;

/**
 * - **Environment: client**
 * 
 * Plays a spatial sound at the position of a specific entity using the specified asset and options.
 * 
 * ```ts
 * // play a sound at the position of a specific entity
 * J.playSoundAtEntity(J.assets.audio.Whistle.id, entityId, {
 *   volume: 0.5,
 *   loop: false,
 * });
 * ```
*/
export const playSoundAtEntity: (
	audioAssetId: string,
	entityId: number,
	options?: { volume?: number; detune?: number; fadeInDuration?: number; loop?: boolean },
) => number | undefined = impl.playSoundAtEntity;

/**
 * - **Environment: client**
 * 
 * Sets the volume of a playing sound, optionally fading to it over `fadeDuration` seconds.
 * 
 * ```ts
 * const musicId = J.playSound(J.assets.audio.BackgroundMusic.id, { loop: true, fadeInDuration: 2 });
 * 
 * // duck the music over 1.5 seconds, e.g. while a dialog plays
 * J.setSoundVolume(musicId, 0.1, 1.5);
 * ```
*/
export const setSoundVolume: (soundId: number, volume: number, fadeDuration?: number) => void =
	impl.setSoundVolume;

/**
 * - **Environment: client**
 * 
 * Applies a lowpass or highpass filter to a playing sound, sweeping the cutoff frequency to `frequency` (in Hz) over `fadeDuration` seconds. Calling it again retargets the sweep, so filters can transition as the game plays.
 * 
 * ```ts
 * // muffle the music while the player is underwater
 * J.setSoundFilter(musicId, "lowpass", 400, 0.5);
 * 
 * // back to (almost) unfiltered
 * J.setSoundFilter(musicId, "lowpass", 24000, 0.5);
 * ```
*/
export const setSoundFilter: (
	soundId: number,
	type: "lowpass" | "highpass",
	frequency: number,
	fadeDuration?: number,
) => void = impl.setSoundFilter;

/**
 * - **Environment: client**
 * 
 * Removes the filter from a playing sound immediately. For a smooth exit, sweep the filter to an inaudible frequency first (24000 Hz for lowpass, 10 Hz for highpass).
*/
export const clearSoundFilter: (soundId: number) => void = impl.clearSoundFilter;

/**
 * - **Environment: client**
 * 
 * Stops a playing sound.
*/
export const stopSound: (soundId: number) => void = impl.stopSound;

/**
 * - **Environment: server**
 * 
 * Resets the world to it's initial state, and triggers `onPlayerJoin` for all players.
*/
export const restartGame: () => void = impl.restartGame;

/**
 * - **Environment: both**
 * 
 * Sends a chat message to all players or specific players. If `toPlayerIds` is not provided, the message is sent to all players.
 * 
 * ```ts
 * // broadcast a message to all players when a player joins
 * J.onPlayerJoin((playerId) => {
 *   const username = J.getPlayerUsername(playerId)!;
 *   J.sendChatMessage(`Player ${username} has joined the game!`, "#00FF00");
 * });
 * 
 * // send a private message to a specific player
 * J.onPlayerJoin((playerId) => {
 *   J.sendChatMessage(`Psst, welcome to the game!`, "#0000FF", [playerId]);
 * });
 * 
 * // send a global message to all players
 * J.sendChatMessage("The game has started! Good luck everyone!", "#FFFFFF");
 * ```
*/
export const sendChatMessage: (message: string, textColor?: string, toPlayerIds?: EntityId[]) => void =
	impl.sendChatMessage;

/**
 * - **Environment: both**
 * 
 * Returns an array of entity IDs for all players currently connected to the world.
 * 
 * ```ts
 * // get list of all players and send them a message
 * const players = J.getAllPlayers();
 * players.forEach((playerId) => {
 *   const username = J.getPlayerUsername(playerId);
 *   J.sendChatMessage(`Hello ${username}!`, "#00FF00", [playerId]);
 * });
 * ```
*/
export const getAllPlayers: () => EntityId[] = impl.getAllPlayers;

/**
 * Returns an array of entity IDs for all NPCs in the world.
 * 
 * ```ts
 * // make all NPCs look at a specific position
 * const npcs = J.getAllNPCs();
 * npcs.forEach((npcId) => {
 *   J.entityLookAt(npcId, [0, 0, 0]); // look at world origin
 * });
 * ```
*/
export const getAllNPCs: () => EntityId[] = impl.getAllNPCs;

/**
 * - **Environment: both**
 * 
 * Returns an array of entity IDs for all props in the world.
*/
export const getAllProps: () => EntityId[] = impl.getAllProps;

/**
 * - **Environment: both**
 * 
 * Returns an array of entity IDs for all characters in the world.
*/
export const getAllCharacters: () => EntityId[] = impl.getAllCharacters;

/**
 * - **Environment: both**
 * 
 * Returns an array of entity IDs for all entities in the world, including players, NPCs, props, areas, and particle systems.
*/
export const getAllEntities: () => EntityId[] = impl.getAllEntities;

/**
 * - **Environment: client**
 * 
 * Returns an array of entity IDs for all particle systems in the world.
*/
export const getAllParticleSystems: () => EntityId[] = impl.getAllParticleSystems;

/**
 * - **Environment: both**
 * 
 * Returns an array of entity IDs for all areas in the world.
*/
export const getAllAreas: () => EntityId[] = impl.getAllAreas;

/**
 * - **Environment: both**
 * 
 * Returns the creator-defined world options that were set at room creation time. Returns a copy of the options, so mutations don't affect the world. Use this when you need to read world options outside of `onGameStart`.
 * 
 * ```ts
 * J.onGameTick(() => {
 *   const options = J.getWorldOptions();
 *   if (options.gameMode === "deathmatch") {
 *     // ...
 *   }
 * });
 * ```
*/
export const getWorldOptions: () => Record<string, string | number | boolean> = impl.getWorldOptions;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when the game starts. The handler receives the creator-defined world options that were set at room creation time.
*/
export const onGameStart: (fn: (worldOptions: Record<string, string | number | boolean>) => void) => void =
	impl.onGameStart;

/**
 * - **Environment: both**
 * 
 * Registers a handler before each physics tick.
 * 
 * You can use this to read inputs, and modify entity positions and velocities before the physics simulation runs, so the changes take effect in the current tick.
*/
export const onBeforePhysicsTick: (fn: (deltaTime: number, time: number) => void) => void =
	impl.onBeforePhysicsTick;

/**
 * - **Environment: both**
 * 
 * Registers a handler for each game tick.
*/
export const onGameTick: (fn: (deltaTime: number, time: number) => void) => void = impl.onGameTick;

/**
 * - **Environment: client**
 * 
 * Registers a handler for each game render. This only runs on the client side. Runs at the client's render framerate and receives the render-frame `deltaTime` and world `time`.
 * 
 * This is typically used with `J.getScreenPosition` to position UI relative to the 3D world at the same FPS as the game's rendering.
*/
export const onGameRender: (fn: (deltaTime: number, time: number) => void) => void = impl.onGameRender;

/**
 * - **Environment: client**
 * 
 * Registers a handler for drawing 2D overlay UI (HUDs, debug text, minimaps) with the canvas 2D API. The handler runs once per render frame, after the 3D scene has rendered, and receives a `CanvasRenderingContext2D` for an engine-managed 2D canvas. After all handlers have drawn, the engine composites the canvas into the main game canvas, so the overlay is part of the final visible frame and is included in screenshots and recordings of the game canvas.
 * 
 * You draw in CSS pixels: `width` / `height` are the logical canvas size, and the engine handles device pixel ratio for you — text and shapes are always rendered at native resolution, no `devicePixelRatio` handling needed. This is the same coordinate space as `getScreenPosition` (draw at its output directly) and `getPointer` (hit-test against its `x`/`y` directly).
 * 
 * The canvas is cleared by the engine at the start of every frame, so draw everything you need each frame. Registering multiple handlers is fine — they all draw into the same canvas, in registration order.
 * 
 * ```ts
 * J.onGameUiCanvas((ctx, width, height) => {
 *   ctx.fillStyle = "white";
 *   ctx.font = "24px sans-serif";
 *   ctx.fillText(`Score: ${score}`, 32, height - 32);
 * });
 * ```
*/
export const onGameUiCanvas: (
	fn: (
		ctx: CanvasRenderingContext2D,
		width: number,
		height: number,
		deltaTime: number,
		time: number,
	) => void,
) => void = impl.onGameUiCanvas;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a player joins.
*/
export const onPlayerJoin: (
	fn: (
		playerId: number,
		persistentData: JsonSerialisable | undefined,
		playerOptions: Record<string, string | number | boolean>,
	) => void,
) => void = impl.onPlayerJoin;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a player leaves.
*/
export const onPlayerLeave: (fn: (playerId: number) => void) => void = impl.onPlayerLeave;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a character uses an item or ability. `primary` / `secondary` correspond to the two use actions (typically left / right click).  
 * 
 * ```ts
 * // Fires for every character on every use
 * J.onCharacterUse({}, (characterId, primary, secondary) => {
 *   if (primary) console.log("primary used by", characterId);
 *   if (secondary) console.log("secondary used by", characterId);
 * });
 * ```
*/
export const onCharacterUse: (
	traitFilters: OnCharacterUseTraitFilters,
	fn: (characterId: EntityId, primary: boolean, secondary: boolean) => void,
) => void = impl.onCharacterUse;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a character stops using an item or ability.
*/
export const onCharacterUnuse: (
	traitFilters: OnCharacterUseTraitFilters,
	fn: (characterId: EntityId) => void,
) => void = impl.onCharacterUnuse;

/**
 * - **Environment: both**
 * 
 * Registers a handler for chat messages.
*/
export const onChat: (fn: (playerId: EntityId, message: string) => void) => void = impl.onChat;

/**
 * - **Environment: both**
 * 
 * Registers a handler for control press events.
*/
export const onControlPress: (
	control: string,
	fn: (playerId: EntityId, combo: string, press: string) => void,
) => void = impl.onControlPress;

/**
 * - **Environment: both**
 * 
 * Registers a handler for control release events.
*/
export const onControlRelease: (
	control: string,
	fn: (playerId: EntityId, combo: string, press: string) => void,
) => void = impl.onControlRelease;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a block is spawned.
*/
export const onBlockSpawn: (blockTypeTraits: TraitId | TraitId[], fn: (position: Vec3) => void) => void =
	impl.onBlockSpawn;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a block is broken.
*/
export const onBlockBreak: (
	traitFilters: TargetTraitFilters,
	fn: (position: Vec3, characterId: EntityId | undefined) => void,
) => void = impl.onBlockBreak;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a block is built.
*/
export const onBlockBuild: (
	traitFilters: TargetTraitFilters,
	fn: (position: Vec3, characterId: EntityId | undefined) => void,
) => void = impl.onBlockBuild;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a block is sculpted.
*/
export const onBlockSculpt: (
	traitFilters: TargetTraitFilters,
	fn: (position: Vec3, characterId: EntityId | undefined) => void,
) => void = impl.onBlockSculpt;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a block is interacted with.
*/
export const onInteractWithBlock: (
	traitFilters: TargetTraitFilters,
	fn: (position: Vec3, triggeredBy: EntityId) => void,
) => void = impl.onInteractWithBlock;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when an entity is interacted with.
*/
export const onInteractWithEntity: (
	traitFilters: TargetTraitFilters,
	fn: (entity: EntityId, triggeredBy: EntityId) => void,
) => void = impl.onInteractWithEntity;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a projectile hits a block.
*/
export const onProjectileHitBlock: (
	traitFilters: ProjectileTraitFilters,
	fn: (
		shooterId: EntityId | undefined,
		position: Vec3,
		hitPosition: Vec3,
		hitNormal: Vec3,
		distance: number,
	) => void,
) => void = impl.onProjectileHitBlock;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when a projectile hits an entity.
*/
export const onProjectileHitEntity: (
	traitFilters: ProjectileTraitFilters,
	fn: (
		entityId: EntityId,
		shooterId: EntityId | undefined,
		distance: number,
		crit: boolean,
		hitPosition: Vec3,
		hitNormal: Vec3,
	) => void,
) => void = impl.onProjectileHitEntity;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when an entity collision starts.
*/
export const onEntityCollisionStart: (
	traitFilters: CollisionTraitFilters,
	fn: (sourceId: EntityId, targetId: EntityId, contactNormal: Vec3) => void,
) => void = impl.onEntityCollisionStart;

/**
 * - **Environment: both**
 * 
 * Registers a handler for ongoing entity collisions.
*/
export const onEntityCollisionPersisted: (
	traitFilters: CollisionTraitFilters,
	fn: (sourceId: EntityId, targetId: EntityId, contactNormal: Vec3) => void,
) => void = impl.onEntityCollisionPersisted;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when an entity collision ends.
*/
export const onEntityCollisionEnd: (
	traitFilters: CollisionTraitFilters,
	fn: (sourceId: EntityId, targetId: EntityId) => void,
) => void = impl.onEntityCollisionEnd;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when an entity starts colliding with a block.
*/
export const onBlockCollisionStart: (
	traitFilters: CollisionTraitFilters,
	fn: (position: Vec3, normal: Vec3, blockType: string, entityId: EntityId) => void,
) => void = impl.onBlockCollisionStart;

/**
 * - **Environment: both**
 * 
 * Registers a handler for when an entity stops colliding with a block.
*/
export const onBlockCollisionEnd: (
	traitFilters: CollisionTraitFilters,
	fn: (position: Vec3, normal: Vec3, blockType: string, entityId: EntityId) => void,
) => void = impl.onBlockCollisionEnd;

/**
 * - **Environment: both**
 * 
 * Constraints two entities bodies together in a fixed way.
*/
export const createFixedConstraint: (entityA: EntityId, entityB: EntityId) => number =
	impl.createFixedConstraint;

/**
 * - **Environment: both**
 * 
 * Applies a continuous force to an entity at a specific point. Unlike impulses, forces are applied continuously over time. Use this for constant effects like gravity wells, wind, or pushing objects.
 * 
 * ```ts
 * // Apply upward force at the center of a prop
 * J.addEntityForce(propId, [0, 50, 0], [0, 0, 0]);
 * 
 * // Apply force at an offset point to create torque
 * J.addEntityForce(propId, [100, 0, 0], [0, 2, 0]);
 * 
 * ```
*/
export const addEntityForce: (entityId: EntityId, force: Vec3, point: Vec3) => void = impl.addEntityForce;

/**
 * - **Environment: both**
 * 
 * Applies an impulse to the entity. Supports props.
*/
export const addEntityImpulse: (entityId: EntityId, impulse: Vec3) => void = impl.addEntityImpulse;

/**
 * - **Environment: both**
 * 
 * Applies an impulse to the entity at a specific point. Supports props.
*/
export const addEntityImpulseAtPoint: (entityId: EntityId, impulse: Vec3, point: Vec3) => void =
	impl.addEntityImpulseAtPoint;

/**
 * - **Environment: both**
 * 
 * Applies an angular impulse to the entity. Supports props.
*/
export const addEntityAngularImpulse: (entityId: EntityId, angularImpulse: Vec3) => void =
	impl.addEntityAngularImpulse;

/**
 * - **Environment: both**
 * 
 * Gets the velocity of the entity at a specific point in world space. Supports props.
*/
export const getEntityVelocityAtPoint: (entityId: EntityId, point: Vec3) => Vec3 | undefined =
	impl.getEntityVelocityAtPoint;

/**
 * - **Environment: both**
 * 
 * Gets the inverse inertia tensor matrix of an entity. This represents how resistant the entity is to rotational acceleration. Returns undefined if the entity doesn't exist or doesn't have physics.
 * 
 * ```ts
 * const inertia = J.getEntityInverseInertia(propId);
 * if (inertia) {
 *   // Use the inertia matrix for physics calculations
 * }
 * ```
*/
export const getEntityInverseInertia: (entityId: EntityId) => Mat4 | undefined = impl.getEntityInverseInertia;

/**
 * - **Environment: both**
 * 
 * Gets whether the entity's physics body is currently active. Supports props.
 * 
 * An entity may become "inactive" or "go to sleep" if it has been stationary for a while.
 * 
 * Usually applying an impulse, or changing the entity's velocity will automatically reactivate the body. In rare cases you may want to manually activate the body using `activateEntityBody` however.
*/
export const getEntityBodyActive: (entityId: EntityId) => boolean | undefined = impl.getEntityBodyActive;

/**
 * - **Environment: both**
 * 
 * Activates the entity's physics body. Supports props.
*/
export const activateEntityBody: (entityId: EntityId) => void = impl.activateEntityBody;

/**
 * - **Environment: both**
 * 
 * Gets the mass of the entity. Supports characters and props.
*/
export const getEntityMass: (entityId: EntityId) => number | undefined = impl.getEntityMass;

/**
 * - **Environment: both**
 * 
 * Constrains a prop entity's movement by setting which translation and rotation axes are allowed. Only works on prop entities.
 * 
 * ```ts
 * const propId = J.spawnProp("prop#ABCDE12345");
 * // movement X, Y, Z, rotation X, Y, Z
 * // only allow movement along Y axis
 * J.setEntityDegreesOfFreedom(propId, false, true, false, false, false, false);
 * ```
*/
export const setEntityDegreesOfFreedom: (
	entityId: EntityId,
	translationX: boolean,
	translationY: boolean,
	translationZ: boolean,
	rotationX: boolean,
	rotationY: boolean,
	rotationZ: boolean,
) => void = impl.setEntityDegreesOfFreedom;

/**
 * - **Environment: both**
 * 
 * Spawns a projectile with the specified properties. The projectile will be fired from the origin position in the given direction.
 * 
 * ```ts
 * const viewRay = J.getCharacterViewRay(playerId);
 * if (viewRay) {
 *   J.spawnProjectile(
 *     "#FF0000",
 *     20,
 *     playerId,
 *     null,
 *     viewRay.origin,
 *     viewRay.direction
 *   );
 * }
 * ```
*/
export const spawnProjectile: (
	color: string,
	velocity: number,
	shooterId: EntityId,
	itemAssetId: string | null,
	origin: Vec3,
	direction: Vec3,
) => void = impl.spawnProjectile;

/**
 * - **Environment: both**
 * 
 * Spawns a 3D text entity at the specified position with the given properties.
 * 
 * ```ts
 * const textId = J.spawnText3D("Hello World!", 32, 50, true, [0, 5, 0], [0, 0, 0]);
 * ```
*/
export const spawnText3D: (
	text: string,
	fontSize: number,
	maxVisibleDistance: number,
	faceCamera: boolean,
	position: Vec3,
	rotation: Vec3,
) => EntityId = impl.spawnText3D;

/**
 * - **Environment: both**
 * 
 * Updates the properties of an existing Text3D entity. All parameters except entityId are optional.
 * 
 * ```ts
 * // update text and size
 * J.updateText3D(textId, "Updated text!", 48);
 * // update text position
 * J.updateText3D(textId, undefined, undefined, [10, 5, 10]);
 * ```
*/
export const updateText3D: (
	entityId: EntityId,
	text?: string,
	fontSize?: number,
	position?: Vec3,
	rotation?: Vec3,
) => void = impl.updateText3D;

/**
 * - **Environment: both**
 * 
 * Finds a path from the start position to the goal position using a voxel grid-based search. The searchSize determines how many search iterations to perform, with larger values allowing for more complex paths but taking longer to compute.
 * 
 * ```ts
 * const start: J.Vec3 = [0, 0, 0];
 * const goal: J.Vec3 = [10, 0, 10];
 * const maxSearchSize = 100;
 * const result = J.findPath(start, goal, maxSearchSize);
 * 
 * // console.log(result.success, result.path);
 * ```
*/
export const findPath: (
	start: Vec3,
	goal: Vec3,
	searchSize: number,
) => {
	success: boolean;
	path: Vec3[];
} = impl.findPath;

/**
 * - **Environment: both**
 * 
 * Finds all walkable blocks within a specified radius from the start position. The searchSize determines how many search iterations to perform, with larger values allowing for more complex searches but taking longer to compute.
 * 
 * ```ts
 * const start: J.Vec3 = [0, 0, 0];
 * const radius = 10;
 * const maxSearchSize = 100;
 * const walkableBlocks = J.findWalkableBlocksInRadius(start, radius, maxSearchSize);
 * 
 * console.log("Walkable blocks:", walkableBlocks);
 * ```
*/
export const findWalkableBlocksInRadius: (start: Vec3, radius: number, searchSize: number) => Vec3[] =
	impl.findWalkableBlocksInRadius;

/**
 * - **Environment: both**
 * 
 * Makes a character entity use an item or perform an action. The `primary` and `secondary` parameters indicate which action to perform.
 * 
 * ```ts
 * J.onGameStart(() => {
 *   // spawn a character
 *   const characterId = J.spawnCharacter(J.assets.avatars.MyAvatar.id);
 * 
 *   // make the character perform a primary use action
 *   J.characterUse(characterId, true, false);
 * 
 *   // make the character perform a secondary use action
 *   J.characterUse(characterId, false, true);
 * });
 * ```
*/
export const characterUse: (id: EntityId, primary: boolean, secondary: boolean) => void = impl.characterUse;

/**
 * - **Environment: both**
 * 
 * Makes a character entity jump with the specified velocity. Only works on character entities (players and NPCs).
 * 
 * ```ts
 * // make all players jump with high velocity
 * const players = J.getAllPlayers();
 * players.forEach((playerId) => {
 *   J.characterJump(playerId, 15.0, true, false);
 * });
 * ```
*/
export const characterJump: (
	characterId: EntityId,
	velocity: number,
	force: boolean,
	predictable: boolean,
) => void = impl.characterJump;

/**
 * - **Environment: both**
 * 
 * Returns whether the character is currently crouching.
*/
export const getCharacterCrouching: (characterId: EntityId) => boolean = impl.getCharacterCrouching;

/**
 * - **Environment: both**
 * 
 * Sets whether the character is crouching. This only works on NPC character entities.
*/
export const setCharacterCrouching: (characterId: EntityId, crouching: boolean) => void =
	impl.setCharacterCrouching;

/**
 * - **Environment: both**
 * 
 * Gets whether a character is currently sprinting (running faster than normal).
 * 
 * ```ts
 * const isSprinting = J.getCharacterSprinting(playerId);
 * if (isSprinting) {
 *   // Apply stamina drain
 *   drainStamina(playerId);
 * }
 * ```
*/
export const getCharacterSprinting: (characterId: EntityId) => boolean = impl.getCharacterSprinting;

/**
 * - **Environment: both**
 * 
 * Sets whether a character should be sprinting. Use this for custom sprint mechanics or forced movement states.
 * 
 * ```ts
 * // Force character to sprint
 * J.setCharacterSprinting(playerId, true);
 * 
 * // Stop character from sprinting
 * J.setCharacterSprinting(playerId, false);
 * ```
*/
export const setCharacterSprinting: (characterId: EntityId, sprinting: boolean) => void =
	impl.setCharacterSprinting;

/**
 * - **Environment: both**
 * 
 * Sets the movement targets for a character entity. The character will move towards these targets in order.
 * 
 * ```ts
 * J.onGameStart(() => {
 *   // spawn a character
 *   const characterId = J.spawnCharacter(J.assets.avatars.MyAvatar.id);
 * 
 *   // set movement targets for the character
 *   const targets: J.Vec3[] = [
 *     [10, 0, 10],
 *     [20, 0, 20],
 *     [30, 0, 30],
 *   ];
 *   J.setCharacterMoveTarget(characterId, targets);
 * });
 * ```
*/
export const setCharacterMoveTarget: (id: EntityId, targets: Vec3[]) => void = impl.setCharacterMoveTarget;

/**
 * - **Environment: both**
 * 
 * Clears the movement targets for a character entity. The character will stop moving towards any targets.
 * 
 * ```ts
 * // clear the movement targets for a character
 * J.clearCharacterMoveTarget(characterId);
 * ```
*/
export const clearCharacterMoveTarget: (id: EntityId) => void = impl.clearCharacterMoveTarget;

/**
 * - **Environment: both**
 * 
 * Sets the "look at" target for a character.
 * 
 * If a character also has a "move target" it will continue looking at the "look at target" while moving to the next "move target".
*/
export const setCharacterLookAtTarget: (id: EntityId, target: Vec3) => void = impl.setCharacterLookAtTarget;

/**
 * - **Environment: both**
 * 
 * Clears the "look at" target for a character.
*/
export const clearCharacterLookAtTarget: (id: EntityId) => void = impl.clearCharacterLookAtTarget;

/**
 * - **Environment: server**
 * 
 * Gets the persistent data associated with a player. This data persists across game sessions and can be used to store player progress, preferences, or other persistent state.
 * 
 * Note that play mode and edit mode have separate persistent data storage - data saved in play mode is not accessible in edit mode and vice versa.
*/
export const getPlayerPersistentData: (playerId: EntityId) => JsonSerialisable | undefined =
	impl.getPlayerPersistentData;

/**
 * - **Environment: server**
 * 
 * Sets the persistent data associated with a player. This data persists across game sessions and can be used to store player progress, preferences, or other persistent state.
 * 
 * Note that play mode and edit mode have separate persistent data storage - data saved in play mode is not accessible in edit mode and vice versa.
 * 
 * ```ts
 * J.onPlayerJoin((playerId, data) => {
 *   // If player doesn't have anything, add 100 coins
 *   if (data === undefined) {
 *     data = { coins: 100 };
 *     J.setPlayerPersistentData(playerId, data);
 *   }
 * 
 *   J.sendChatMessage(
 *     `Player ${playerId} has joined with ${data.coins} coins`,
 *     "#00FF00",
 *   );
 * });
 * 
 * // give players 3 coins per 1 second
 * const INTERVAL = 1;
 * const COINS_PER_INTERVAL = 3;
 * 
 * let delay = INTERVAL;
 * 
 * J.onGameTick((dt) => {
 *   delay -= dt;
 * 
 *   if (delay > 0) return;
 * 
 *   delay = INTERVAL;
 * 
 *   const players = J.getAllPlayers();
 *   players.forEach((playerId) => {
 *     const data = J.getPlayerPersistentData(playerId);
 * 
 *     if (data !== undefined) {
 *       data.coins += COINS_PER_INTERVAL;
 *       J.setPlayerPersistentData(playerId, data);
 *       J.sendChatMessage(`You have ${data.coins} coins`, "#00FF00", [playerId]);
 *     }
 *   });
 * });
 * ```
*/
export const setPlayerPersistentData: (playerId: EntityId, data: JsonSerialisable) => void =
	impl.setPlayerPersistentData;

/**
 * - **Environment: client**
 * 
 * Converts a 3D world position to a 2D screen position. Returns the screen coordinates where the world position would appear on the player's screen.
 * 
 * The returned Vec2 contains x and y coordinates in screen space, where (0,0) is typically the top-left corner of the screen. If the world position is behind the camera or outside the view frustum, the returned coordinates may be outside the visible screen area.
*/
export const getScreenPosition: (out: Vec2, worldPosition: Vec3) => Vec2 = impl.getScreenPosition;

/**
 * - **Environment: both**
 * 
 * Sets the current environment preset for the world using the specified asset ID. This will change lighting, fog, sky, and other environmental settings.
 * Environment presets can be created and managed through the world editor. It is recommended to avoid changing this frequently.
 * 
 * ```ts
 * // Day time
 * J.setEnvironmentPreset(J.assets.environmentPresets.DayPreset.id);
 * // Night time
 * J.setEnvironmentPreset(J.assets.environmentPresets.NightPreset.id);
 * ```
*/
export const setEnvironmentPreset: (assetId: string) => void = impl.setEnvironmentPreset;

/**
 * - **Environment: client**
 * 
 * Client-only function that presents a world portal to the player.
 * 
 * ```ts
 * J.showWorldPortal("XIMXA");
 * ```
*/
export const showWorldPortal: (worldShortcode: string) => void = impl.showWorldPortal;

export const join: (
	worldOptions?: Record<string, string | number | boolean>,
	playerOptions?: Record<string, string | number | boolean>,
) => void = impl.join;

/**
 * - **Environment: client**
 * 
 * Returns the prop's asset icon as a base64-encoded string url, or undefined if no icon is available. This can be used to display visual representations of props in UI elements.
*/
export const getPropAssetIcon: (assetId: string) => string | undefined = impl.getPropAssetIcon;

export const trackEvent: (event: string, value: number, playerId?: number) => void = impl.trackEvent;

/**
 * - **Environment: client**
 * 
 * Returns the current camera position, quaternion, vertical field of view in degrees, and aspect ratio (width / height of the render viewport).
*/
export const getCamera: () => {
	position: Vec3;
	quaternion: Quat;
	fov: number;
	aspect: number; // width / height of the render viewport
} = impl.getCamera;

/**
 * - **Environment: client**
 * 
 * Sets the camera to a free-floating position with custom rotation, detached from the local player. This allows you to create cinematic camera movements, security cameras, or other custom camera behaviors.
 * 
 * - `position` - The 3D world position where the camera should be placed
 * - `rotation` - The quaternion rotation for the camera orientation
 * 
 * ```ts
 * // Set a fixed camera position looking at a specific area
 * J.setCameraFree([10, 5, 10], [0, 0, 0, 1]);
 * 
 * // Create a camera that follows an entity
 * const cameraOffset = J.vec3.fromValues(0, 4, 10);
 * const cameraLocalMatrix = J.mat4.fromTranslation(J.mat4.create(), cameraOffset);
 * const rot = J.quat.create();
 * const pos = J.vec3.create();
 * const _scale = J.vec3.create();
 * const cameraWorldMatrix = J.mat4.create();
 * 
 * J.onGameRender(() => {
 *     const localPlayerId = J.getLocalPlayer();
 *     const mat = J.getEntityMatrix(localPlayerId);
 * 
 *     if (mat) {
 *         J.mat4.multiply(cameraWorldMatrix, mat, cameraLocalMatrix);
 *         J.mat4.decompose(rot, pos, _scale, cameraWorldMatrix);
 *         J.setCameraFree(pos, rot);
 *     }
 * });
 * ```
*/
export const setCameraFree: (position: Vec3, rotation: Quat) => void = impl.setCameraFree;

/**
 * - **Environment: client**
 * 
 * Sets the camera to the local player with specific allowed camera modes. Can be used both at game start to configure initial camera modes, or to reattach the camera after using setCameraFree.
 * 
 * - `allowedModes` - Array of camera modes that the player can switch between
 * 
 * ```ts
 * // Return to normal player camera after using free camera
 * J.setLocalPlayerCamera(["firstPerson", "thirdPerson", "selfie"]);
 * 
 * // Restrict to first-person only (no camera switching)
 * J.setLocalPlayerCamera(["firstPerson"]);
 * 
 * // Allow first-person and third-person, but not selfie mode
 * J.setLocalPlayerCamera(["firstPerson", "thirdPerson"]);
 * 
 * // Set up camera modes when game starts
 * J.onGameStart(() => {
 *   // Configure camera to only allow first-person view
 *   J.setLocalPlayerCamera(["firstPerson"]);
 * });
 * ```
*/
export const setLocalPlayerCamera: (allowedModes: ("firstPerson" | "thirdPerson" | "selfie")[]) => void =
	impl.setLocalPlayerCamera;

/**
 * - **Environment: both**
 * 
 * Defines a new leaderboard with the specified ID and scoring mode. This should typically be called in a shared file so both client and server have access to the same leaderboard definition.
 * 
 * - `leaderboardId` - A unique string identifier for the leaderboard
 * - `options.mode` - Either "lowest" (for time-based competitions) or "highest" (for score-based competitions)
 * 
 * ```ts
 * // shared.ts - Define leaderboards that both client and server can access
 * export const RaceLeaderboard = J.leaderboards.defineLeaderboard("race_times", { 
 *   mode: "lowest" 
 * });
 * 
 * export const ScoreLeaderboard = J.leaderboards.defineLeaderboard("high_scores", { 
 *   mode: "highest" 
 * });
 * ```
*/
export const leaderboards: {
	defineLeaderboard: (leaderboardId: string, options: { mode: "lowest" | "highest" }) => LeaderboardId;
	getPlayerScore: (leaderboard: LeaderboardId, playerId: EntityId) => Promise<LeaderboardEntry | null>;
	setPlayerScore: (leaderboard: LeaderboardId, playerId: EntityId, score: number) => void;
	getTopScores: (leaderboard: LeaderboardId) => Promise<LeaderboardEntry[]>;
	getScoresForPlayers: (leaderboard: LeaderboardId) => Promise<LeaderboardEntry[]>;
} = impl.leaderboards;
