// jamango/avatars.ts
import { CHARACTER_RIG_IDS } from "./character-component";
import type { CharacterRigId, GlbAnimationClipName, MeshComponentId, MeshSetId } from "./character-component";

// this file must stay self-contained (only ./character-component imports) — it ships
// verbatim into the creator-facing `jamango` package, see buildJamangoPackageFiles
export { CHARACTER_RIG_IDS };
export type { CharacterRigId, GlbAnimationClipName, MeshComponentId, MeshSetId };

// ---------------------------------------------------------------------------
// Character avatar runtime constants & types
// ---------------------------------------------------------------------------

export const CHARACTER_SLOT_IDS = [
	"head",
	"chest",
	"waist",
	"armLeft",
	"armRight",
	"handLeft",
	"handRight",
	"legLeft",
	"legRight",
	"footLeft",
	"footRight",
] as const;

export type CharacterSlotId = (typeof CHARACTER_SLOT_IDS)[number];

/** A character component id. Format is `${setId}_${slotKey}_${partName}` for body parts and `item_${partName}` for items.
 * `MeshComponentId | (string & {})` surfaces autocomplete on the built-in ids while still accepting any string
 * (third-party/product ids), so structure is still looked up defensively via MESH_COMPONENTS. */
export type CharacterComponentId = MeshComponentId | (string & {});

export type CharacterAvatarFaceConfig = {
	eyeId: string | null;
	pupilId: string | null;
	mouthId: string | null;
	browId: string | null;
	featureId: string | null;
	eyeXOffset: number;
	eyeYOffset: number;
	zOffset: number;
	eyeScale: number;
	mouthYOffset: number;
	mouthScale: number;
	featureXOffset: number;
	featureYOffset: number;
	featureScale: number;
	browXOffset: number;
	browYOffset: number;
	browScale: number;
	browZRotation: number;
};

export type CharacterAvatarAnimKey = "idle" | "run" | "walk" | "crouch";

export type CharacterAvatarAnimConfig = Record<CharacterAvatarAnimKey, GlbAnimationClipName | null>;

/** Palette slot keys exposed to scripters. Matches the colorKey values in MESH_MATERIAL_ENTRIES. */
export type CharacterAvatarComponentColorKey = "01" | "02" | "03" | "04" | "05";

/** Color overrides for a component. Keys autocomplete to the five palette slots. */
export type CharacterAvatarComponentColors = Partial<Record<CharacterAvatarComponentColorKey, string>>;

export type CharacterAvatarComponentConfig = {
	id: CharacterComponentId;
	colors?: CharacterAvatarComponentColors;
	/** Baseline default component supplied by createDefaultCharacterAvatarConfig.
	 * Dropped from a slot when a replacement overlay contributes to that slot; mix
	 * components stack with it — see mergeCharacterAvatarConfig / assembleCharacterAvatarConfig. */
	isFallback?: boolean;
};

export type CharacterAvatarConfig = {
	rigId: CharacterRigId | null;
	skinColorPrimary: string | null;
	skinColorSecondary: string | null;
	hairColor: string | null;
	anim: CharacterAvatarAnimConfig;
	face: CharacterAvatarFaceConfig;
	components: Record<CharacterSlotId, CharacterAvatarComponentConfig[]>;
};

/**
 * Exposed mod API type for avatar updates. Scripters pass partials of CharacterAvatarConfig for J.addCharacterAvatarConfig
 */
export type PartialCharacterAvatarConfig = Partial<Omit<CharacterAvatarConfig, "anim" | "face" | "components">> & {
	// makes both nested objects partial to allow partial updates without needing to specify the entire face/components structure
	anim?: Partial<CharacterAvatarAnimConfig>;
	face?: Partial<CharacterAvatarFaceConfig>;
	components?: Partial<Record<CharacterSlotId, CharacterAvatarComponentConfig[]>>;
};
