// jamango/schema.ts
import type { Mat2, Mat3, Mat4, Quat, Vec2, Vec3, Vec4 } from "./math/types";

/* schema type */

// schemas are simple objects

type CommonSchema = {
	label?: string;
	description?: string;

	optional?: boolean;
	nullable?: boolean;

	hidden?: boolean;
};

export type SelectOption<T = any> = {
	value: T;
	label: string;
};

export type BooleanSchema = {
	type: "boolean";
	defaultValue?: boolean;
} & CommonSchema;

export type NumberSchema = {
	type: "number";
	defaultValue?: number;
	control?: "input" | "range" | "select";
	options?: SelectOption<number>[];
	min?: number;
	max?: number;
	step?: number;
	precision?: "integer" | "float";
} & CommonSchema;

export type StringSchema = {
	type: "string";
	defaultValue?: string;
	control?: "input" | "select" | "textarea";
	options?: SelectOption<string>[];
	minLength?: number;
	maxLength?: number;
} & CommonSchema;

export type Vec2Schema = {
	type: "vec2";
	defaultValue?: Vec2;
	control?: "position" | "size" | "vector";
} & CommonSchema;

export type Vec3Schema = {
	type: "vec3";
	defaultValue?: Vec3;
	control?: "position" | "block" | "direction" | "vector";
} & CommonSchema;

export type Vec4Schema = {
	type: "vec4";
	defaultValue?: Vec4;
	control?: "color" | "vector";
} & CommonSchema;

export type Mat2Schema = {
	type: "mat2";
	defaultValue?: Mat2;
} & CommonSchema;

export type Mat3Schema = {
	type: "mat3";
	defaultValue?: Mat3;
} & CommonSchema;

export type Mat4Schema = {
	type: "mat4";
	defaultValue?: Mat4;
} & CommonSchema;

export type EulerSchema = {
	type: "euler";
	defaultValue?: Vec3;
} & CommonSchema;

export type QuatSchema = {
	type: "quat";
	defaultValue?: Quat;
} & CommonSchema;

export type EntityIdSchema = {
	type: "entityId";
	defaultValue?: string;
} & CommonSchema;

export type AssetSchema = {
	type: "asset";
	defaultValue?: string;
	assetTypes?: Array<
		| "prop"
		| "avatar"
		| "block"
		| "audio"
		| "particleSystem"
		| "environmentPreset"
		| "animation"
	>;
} & CommonSchema;

export type TraitSchema = {
	type: "trait";
} & CommonSchema;

export type SceneTreeNodeSchema = {
	type: "sceneTreeNode";
	sceneTreeNodeTypes?: number[];
} & CommonSchema;

export type TraitsSchema = {
	type: "traits";
	targetType?: string;
} & CommonSchema;

export type ListSchema = {
	type: "list";
	of: Schema;
} & CommonSchema;

export type ObjectSchema = {
	type: "object";
	fields: Record<string, Schema>;
} & CommonSchema;

export type RecordSchema = {
	type: "record";
	field: Schema;
} & CommonSchema;

export type AnySchema<T = any> = {
	type: "any";
	__tsType?: T;
} & CommonSchema;

export type LiteralSchema = {
	type: "literal";
	value: any;
} & CommonSchema;

export type UnionSchema<K extends string = string, V extends ObjectSchema[] = ObjectSchema[]> = {
	type: "union";
	key: K;
	variants: [...V];
} & CommonSchema;

export type Schema =
	| BooleanSchema
	| NumberSchema
	| StringSchema
	| Vec2Schema
	| Vec3Schema
	| Vec4Schema
	| Mat2Schema
	| Mat3Schema
	| Mat4Schema
	| EulerSchema
	| QuatSchema
	| EntityIdSchema
	| AssetSchema
	| TraitSchema
	| SceneTreeNodeSchema
	| TraitsSchema
	| ListSchema
	| ObjectSchema
	| RecordSchema
	| AnySchema
	| LiteralSchema
	| UnionSchema;

/* schema to type utility */
type NullableOptionalize<S extends { optional?: boolean; nullable?: boolean }, T> = S["optional"] extends true
	? S["nullable"] extends true
		? T | null | undefined
		: T | undefined
	: S["nullable"] extends true
		? T | null
		: T;

type Simplify<T> = { [K in keyof T]: T[K] } & {};

// maximum recursion depth helper
type NextDepth = {
	0: 0;
	1: 0;
	2: 1;
	3: 2;
	4: 3;
	5: 4;
	6: 5;
	7: 6;
	8: 7;
	9: 8;
	10: 9;
	11: 10;
	12: 11;
	13: 12;
	14: 13;
	15: 14;
};

type DecrementDepth<N extends keyof NextDepth> = N extends keyof NextDepth ? NextDepth[N] : 0;

// biome-ignore format: readability
export type SchemaType<S extends Schema, Depth extends keyof NextDepth = 15> =
	Depth extends 0 ? any :
	S extends BooleanSchema ? NullableOptionalize<S, boolean> :
	S extends NumberSchema ? NullableOptionalize<S, number> :
	S extends StringSchema ? NullableOptionalize<S, string> :
	S extends Vec2Schema ? NullableOptionalize<S, Vec2> :
	S extends Vec3Schema ? NullableOptionalize<S, Vec3> :
	S extends Vec4Schema ? NullableOptionalize<S, Vec4> :
	S extends Mat2Schema ? NullableOptionalize<S, Mat2> :
	S extends Mat3Schema ? NullableOptionalize<S, Mat3> :
	S extends Mat4Schema ? NullableOptionalize<S, Mat4> :
	S extends EulerSchema ? NullableOptionalize<S, Vec3> :
	S extends QuatSchema ? NullableOptionalize<S, Quat> :
	S extends EntityIdSchema ? NullableOptionalize<S, string> :
	S extends AssetSchema ? NullableOptionalize<S, string> :
	S extends TraitSchema ? NullableOptionalize<S, string> :
	S extends SceneTreeNodeSchema ? NullableOptionalize<S, string> :
	S extends TraitsSchema ? NullableOptionalize<S, Record<string, any>> :
	S extends AnySchema<infer T> ? NullableOptionalize<S, T> :
	S extends LiteralSchema ? NullableOptionalize<S, S["value"]> :
	S extends { type: "union"; variants: infer V extends Schema[] } ? NullableOptionalize<S, SchemaType<V[number], DecrementDepth<Depth>>> :
	S extends { type: "list"; of: infer U extends Schema } ? NullableOptionalize<S, SchemaType<U, DecrementDepth<Depth>>[]> :
	S extends { type: "object"; fields: infer F extends Record<string, Schema> } ?
		Simplify<NullableOptionalize<S, ({
			[K in keyof F as F[K] extends Schema ? (F[K]["optional"] extends true ? never : K) : never]:
				F[K] extends Schema ? SchemaType<F[K], DecrementDepth<Depth>> : never
		}) & ({
			[K in keyof F as F[K] extends Schema ? (F[K]["optional"] extends true ? K : never) : never]?:
				F[K] extends Schema ? SchemaType<F[K], DecrementDepth<Depth>> : never
		})>> :
	S extends { type: "record"; field: infer V extends Schema } ? NullableOptionalize<S, Record<string, SchemaType<V, DecrementDepth<Depth>>>> :
	never;

/* lightweight helpers that just return objects */

export const boolean = <O extends Partial<Omit<BooleanSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "boolean" } & O => ({
	...opts,
	type: "boolean",
});

export const number = <O extends Partial<Omit<NumberSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "number" } & O => ({
	...opts,
	type: "number",
});

export const string = <O extends Partial<Omit<StringSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "string" } & O => ({
	...opts,
	type: "string",
});

export const vec2 = <O extends Partial<Omit<Vec2Schema, "type">> = {}>(
	opts: O = {} as O,
): { type: "vec2" } & O => ({
	...opts,
	type: "vec2",
});

export const vec3 = <O extends Partial<Omit<Vec3Schema, "type">> = {}>(
	opts: O = {} as O,
): { type: "vec3" } & O => ({
	...opts,
	type: "vec3",
});

export const vec4 = <O extends Partial<Omit<Vec4Schema, "type">> = {}>(
	opts: O = {} as O,
): { type: "vec4" } & O => ({
	...opts,
	type: "vec4",
});

export const mat2 = <O extends Partial<Omit<Mat2Schema, "type">> = {}>(
	opts: O = {} as O,
): { type: "mat2" } & O => ({
	...opts,
	type: "mat2",
});

export const mat3 = <O extends Partial<Omit<Mat3Schema, "type">> = {}>(
	opts: O = {} as O,
): { type: "mat3" } & O => ({
	...opts,
	type: "mat3",
});

export const mat4 = <O extends Partial<Omit<Mat4Schema, "type">> = {}>(
	opts: O = {} as O,
): { type: "mat4" } & O => ({
	...opts,
	type: "mat4",
});

export const euler = <O extends Partial<Omit<EulerSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "euler" } & O => ({
	...opts,
	type: "euler",
});

export const quat = <O extends Partial<Omit<QuatSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "quat" } & O => ({
	...opts,
	type: "quat",
});

export const asset = <O extends Partial<Omit<AssetSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "asset" } & O => ({
	...opts,
	type: "asset",
});

export const entityId = <O extends Partial<Omit<EntityIdSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "entityId" } & O => ({
	...opts,
	type: "entityId",
});

export const trait = <O extends Partial<Omit<TraitSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "trait" } & O => ({
	...opts,
	type: "trait",
});

export const sceneTreeNode = <O extends Partial<Omit<SceneTreeNodeSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "sceneTreeNode" } & O => ({
	type: "sceneTreeNode",
	...opts,
});

export const traits = <O extends Partial<Omit<TraitsSchema, "type">> = {}>(
	opts: O = {} as O,
): { type: "traits" } & O => ({
	...opts,
	type: "traits",
});

export const any = <T = any, O extends Partial<Omit<AnySchema<T>, "type" | "__tsType">> = {}>(
	opts: O = {} as O,
): AnySchema<T> & O =>
	({
		type: "any",
		...opts,
	}) as AnySchema<T> & O;

/**
 * Literal schema - constant value used for discriminators in unions.
 * The value is not intended to be editable and consumes no serialized bytes.
 */
export const literal = <V extends string | number | boolean, O extends Partial<CommonSchema> = {}>(
	value: V,
	opts: O = {} as O,
) => {
	return {
		type: "literal",
		value,
		...opts,
	} as const;
};

/**
 * Union schema - discriminated union of object variants.
 * key - name of the discriminator field
 * variants - array of object schemas where each variant must include a literal field for the key
 */
export const union = <K extends string, V extends (ObjectSchema & { fields: { [k in K]: LiteralSchema } })[]>(
	key: K,
	variants: [...V],
): { type: "union"; key: K; variants: [...V] } => ({
	type: "union",
	key,
	variants,
});

export function list<T extends Schema>(of: T): { type: "list"; of: T };
export function list<T extends Schema, O extends Partial<Omit<ListSchema, "type" | "of">>>(
	of: T,
	opts: O,
): { type: "list"; of: T } & O;
export function list<T extends Schema, O extends Partial<Omit<ListSchema, "type" | "of">> = {}>(
	of: T,
	opts?: O,
): { type: "list"; of: T } | ({ type: "list"; of: T } & O) {
	return {
		type: "list",
		of,
		...(opts as O | undefined),
	} as unknown as any;
}

export function object<F extends Record<string, Schema>>(fields: F): { type: "object"; fields: F };
export function object<
	F extends Record<string, Schema>,
	O extends Partial<Omit<ObjectSchema, "type" | "fields">>,
>(fields: F, opts: O): { type: "object"; fields: F } & O;
export function object<
	F extends Record<string, Schema>,
	O extends Partial<Omit<ObjectSchema, "type" | "fields">> = {},
>(fields: F, opts?: O): { type: "object"; fields: F } | ({ type: "object"; fields: F } & O) {
	return {
		type: "object",
		fields,
		...(opts as O | undefined),
	} as unknown as any;
}

export function record<F extends Record<string, Schema>>(fields: F): { type: "record"; field: F };
export function record<F extends Schema, O extends Partial<Omit<RecordSchema, "type" | "field">>>(
	field: F,
	opts: O,
): { type: "record"; field: F } & O;
export function record<F extends Schema, O extends Partial<Omit<RecordSchema, "type" | "field">> = {}>(
	field: F,
	opts?: O,
): { type: "record"; field: F } | ({ type: "record"; field: F } & O) {
	return {
		type: "record",
		field,
		...(opts as O | undefined),
	} as unknown as any;
}
