// jamango/types.ts
import * as Api from "./api";
import type { Quat, Vec3 } from "./math/types";
import type { Schema } from "./schema";
import type { TraitMeta } from "./traits";

export type Impl = typeof Api;

export type EntityId = number;

export type CharacterViewRay = {
	origin: Vec3;
	direction: Vec3;
	entity: EntityId | undefined;
	block: Vec3 | undefined;
};

export type CharacterVisualFacingMode = "movement" | "camera";

export type PlayerPermissions = {
	canUseIndividualBlocks: boolean;
	canFly: boolean;
	canInteract: boolean;
	canForceRespawn: boolean;
};

export type CharacterMovementProperties = {
	walkSpeed: number;
	crouchSpeed: number;
	airSpeed: number;
	flySpeed: number;
	jumpVelocity: number;
	terminalVelocity: number;
	sprintSpeedMultiplier: number;
	jumpCooldownTime: number;
	coyoteTime: number;
	slidingGraceTime: number;
	canSprint: boolean;
	canCrouch: boolean;
	canJump: boolean;
	maxHorizontalSpeedMultiplier: number;
	groundAcceleration: number;
	groundFriction: number;
	groundFrictionStopMultiplier: number;
	groundStopSpeed: number;
	airAcceleration: number;
	airFriction: number;
	airFrictionStopMultiplier: number;
	gravityFactor: number;
};

export type MotionType =
	| typeof Api.MOTION_TYPE_DYNAMIC
	| typeof Api.MOTION_TYPE_STATIC
	| typeof Api.MOTION_TYPE_KINEMATIC;

/**
 * What an animation clip does once it reaches its end:
 * - `loop` — restart from the beginning, indefinitely.
 * - `once` — play through once, then clear the clip (base returns to locomotion, layers stop).
 * - `hold` — play through once, then freeze on the final frame until explicitly stopped.
 */
export type AnimationPlaybackMode = "loop" | "once" | "hold";

export type TraitId = string;
export type TraitTargetId = string | number;

export type TraitDefinition = {
	id: string;
	schema?: Schema;
	meta?: TraitMeta;
};

export type OnCharacterUseTraitFilters = {
	character?: TraitId[];
};

export type TargetTraitFilters = {
	target?: TraitId[];
};

export type ProjectileTraitFilters = {
	item?: TraitId[];
	block?: TraitId[];
	shot?: TraitId[];
	shooter?: TraitId[];
};

export type CollisionTraitFilters = {
	source?: TraitId[];
	target?: TraitId[];
};

type JsonSerialisablePrimitive = string | number | boolean | null;
type JsonSerialisableValue = JsonSerialisablePrimitive | JsonSerialisableObject | JsonSerialisableArray;
type JsonSerialisableObject = {
	[key: string]: JsonSerialisableValue;
};
type JsonSerialisableArray = Array<JsonSerialisableValue>;
export type JsonSerialisable = JsonSerialisableValue;

export type TraitType = { _traitDataType: any };

import type {
	CharacterAvatarAnimConfig,
	CharacterAvatarAnimKey,
	CharacterAvatarConfig,
	CharacterAvatarComponentConfig,
	CharacterAvatarComponentColorKey,
	CharacterAvatarComponentColors,
	CharacterAvatarFaceConfig,
	CharacterComponentId,
	CharacterRigId,
	CharacterSlotId,
	MeshComponentId,
	MeshSetId,
	PartialCharacterAvatarConfig,
} from "@jamango/content-client";

export type {
	CharacterAvatarAnimConfig,
	CharacterAvatarAnimKey,
	CharacterAvatarConfig,
	CharacterAvatarComponentConfig,
	CharacterAvatarComponentColorKey,
	CharacterAvatarComponentColors,
	CharacterAvatarFaceConfig,
	CharacterComponentId,
	CharacterRigId,
	CharacterSlotId,
	MeshComponentId,
	MeshSetId,
	PartialCharacterAvatarConfig,
};

// Inline GLB component descriptor. `id` keys the shared `geometryById` map; `colors`
// optionally overrides per-material palette colors ("01".."05"). Client-only.
export type GlbMeshSpec = {
	id: CharacterComponentId;
	colors?: CharacterAvatarComponentColors;
};

// Physics overrides for spawnPropFromSpec, stored on the entity's def.
// TODO: inert until sphere/box collider allocation for non-block-structure props lands.
export type SpecPropPhysics = {
	motionType?: MotionType;
	mass?: number;
	friction?: number;
	restitution?: number;
};

export type CharacterMeshAttachmentInfo = {
	handle: number;
	component: string;
	slot: CharacterSlotId;
	position: Vec3;
	quaternion: Quat;
	scale: Vec3;
};

export type FirstPersonMeshAttachmentInfo = {
	handle: number;
	component: string;
	position: Vec3;
	quaternion: Quat;
	scale: Vec3;
};

export type LeaderboardId = string;
export type LeaderboardEntry = {
	place: number;
	score: number;
	accountId: string;
	playerId?: EntityId;
	playerUsername: string;
	playerAvatar?: string;
};

export type WorldBounds = {
	minX: number;
	minY: number;
	minZ: number;
	maxX: number;
	maxY: number;
	maxZ: number;
};

export type WorldSpawn = {
	position: Vec3;
	angle: number;
};

export type WorldSettings = {
	chatEnabled: boolean;
};

export type UISettings = {
	mobileCameraButton: boolean;
	mobileInteractButton: boolean;
	mobileEmoteButton: boolean;
	mobileRightButtons: "default" | false;
};
